<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Opqua &mdash; Opqua 1.0.2 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="_static/documentation_options.js?v=1ed6394b"></script>
        <script src="_static/doctools.js?v=888ff710"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            Opqua
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="about.html">About</a></li>
<li class="toctree-l1"><a class="reference internal" href="requirements_and_installation.html">Requirements and Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="model_documentation.html">Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="modules.html">API</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Opqua</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Opqua <img align="right" src="../img/circle-header.png" width="400px" title="circle header"></li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/README.md.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="opqua">
<h1>Opqua <img align="right" src="../img/circle-header.png" width="400px" title="circle header"><a class="headerlink" href="#opqua" title="Link to this heading"></a></h1>
<p><a class="reference external" href="https://zenodo.org/badge/latestdoi/249037110"><img alt="DOI" src="https://zenodo.org/badge/249037110.svg" /></a></p>
<p><strong>opqua</strong> (opkua, upkua)
[<a class="reference external" href="https://en.wikipedia.org/wiki/Chibcha_language">Chibcha/muysccubun</a>]</p>
<p><strong>I.</strong> <em>noun</em>. ailment, disease, illness</p>
<p><strong>II.</strong> <em>noun</em>. cause, reason [<em>for which something occurs</em>]</p>
<p><em>Taken from D. F. Gómez Aldana’s
<a class="reference external" href="http://muysca.cubun.org/opqua">muysca-spanish dictionary</a></em>.</p>
<section id="opqua-is-an-epidemiological-modeling-framework-for-pathogen-population-genetics-and-evolution">
<h2>Opqua is an epidemiological modeling framework for pathogen population genetics and evolution.<a class="headerlink" href="#opqua-is-an-epidemiological-modeling-framework-for-pathogen-population-genetics-and-evolution" title="Link to this heading"></a></h2>
<p>Opqua stochastically simulates pathogens with distinct, evolving genotypes that
spread through populations of hosts which can have specific immune profiles.</p>
<p>Opqua is a useful tool to test out scenarios, explore hypotheses, make
predictions, and teach about the relationship between pathogen evolution and
epidemiology.</p>
<p>Among other things, Opqua can model</p>
<ul class="simple">
<li><p>host-host, vector-borne, and vertical transmission</p></li>
<li><p>pathogen evolution through mutation, recombination, and/or reassortment</p></li>
<li><p>host recovery, death, and birth</p></li>
<li><p>metapopulations with complex structure and demographic interactions</p></li>
<li><p>interventions and events altering demographic, ecological, or evolutionary
parameters</p></li>
<li><p>treatment and immunization of hosts or vectors</p></li>
<li><p>influence of pathogen genome sequences on transmission and evolution, as well
as host demographic dynamics</p></li>
<li><p>intra- and inter-host competition and evolution of pathogen strains across
user-specified adaptive landscapes</p></li>
</ul>
<p>Check out the <code class="docutils literal notranslate"><span class="pre">changelog</span></code> file for information on recent updates.</p>
<p>Opqua has been used in-depth to study <a class="reference external" href="https://github.com/pablocarderam/fitness_valleys_opqua">pathogen evolution across fitness valleys</a>.
Check out the peer-reviewed preprint on
<a class="reference external" href="https://doi.org/10.1101/2021.12.16.473045">biorXiv</a>, now peer-reviewed.</p>
<p>Opqua is developed by <a class="reference external" href="https://pablo-cardenas.com">Pablo Cárdenas</a> in
collaboration with Vladimir Corredor and Mauricio Santos-Vega.
Follow their science antics on Twitter at
<a class="reference external" href="https://twitter.com/pcr_guy">&#64;pcr_guy</a> and
<a class="reference external" href="https://twitter.com/msantosvega">&#64;msantosvega</a>.</p>
<p>Opqua is <a class="reference external" href="https://pypi.org/project/opqua/">available on PyPI</a> and is distributed
under an <a class="reference external" href="https://choosealicense.com/licenses/mit/">MIT License</a>.</p>
</section>
<section id="example-plots">
<h2>Example Plots<a class="headerlink" href="#example-plots" title="Link to this heading"></a></h2>
<p>These are some of the plots Opqua is able to produce, but you can output the
raw simulation data yourself to make your own analyses and plots. These are all
taken from the examples in the <code class="docutils literal notranslate"><span class="pre">examples/tutorials</span></code> folder—try them out
yourself! See the
[Requirements and Installation](#Requirements and Installation) and
<a class="reference internal" href="#usage"><span class="std std-ref">Usage</span></a> sections for more details.</p>
<section id="population-genetic-composition-plots-for-pathogens">
<h3>Population genetic composition plots for pathogens<a class="headerlink" href="#population-genetic-composition-plots-for-pathogens" title="Link to this heading"></a></h3>
<p>An optimal pathogen genome arises and outcompetes all others through intra-host
competition. See <code class="docutils literal notranslate"><span class="pre">fitness_function_mutation_example.py</span></code> in the
<code class="docutils literal notranslate"><span class="pre">examples/tutorials/evolution</span></code> folder.
<img alt="Compartments" src="_images/fitness_function_mutation_example_composition.png" /></p>
</section>
<section id="host-vector-compartment-plots">
<h3>Host/vector compartment plots<a class="headerlink" href="#host-vector-compartment-plots" title="Link to this heading"></a></h3>
<p>A population with natural birth and death dynamics shows the effects of a
pathogen. “Dead” denotes deaths caused by pathogen infection. See
<code class="docutils literal notranslate"><span class="pre">vector-borne_birth-death_example.py</span></code> in the <code class="docutils literal notranslate"><span class="pre">examples/tutorials/vital_dynamics</span></code>
folder.
<img alt="Compartments" src="_images/vector-borne_birth-death_example.png" /></p>
</section>
<section id="plots-of-a-host-vector-compartment-across-different-populations-in-a-metapopulation">
<h3>Plots of a host/vector compartment across different populations in a metapopulation<a class="headerlink" href="#plots-of-a-host-vector-compartment-across-different-populations-in-a-metapopulation" title="Link to this heading"></a></h3>
<p>Pathogens spread through a network of interconnected populations of hosts. Lines
denote infected pathogens. See
<code class="docutils literal notranslate"><span class="pre">metapopulations_migration_example.py</span></code> in the
<code class="docutils literal notranslate"><span class="pre">examples/tutorials/metapopulations</span></code> folder.
<img alt="Compartments" src="_images/metapopulations_migration_example.png" /></p>
</section>
<section id="id1">
<h3>Host/vector compartment plots<a class="headerlink" href="#id1" title="Link to this heading"></a></h3>
<p>A population undergoes different interventions, including changes in
epidemiological parameters and vaccination. “Recovered” denotes immunized,
uninfected hosts.
See <code class="docutils literal notranslate"><span class="pre">intervention_examples.py</span></code> in the <code class="docutils literal notranslate"><span class="pre">examples/tutorials/interventions</span></code> folder.
<img alt="Compartments" src="_images/intervention_examples_compartments.png" /></p>
</section>
<section id="pathogen-phylogenies">
<h3>Pathogen phylogenies<a class="headerlink" href="#pathogen-phylogenies" title="Link to this heading"></a></h3>
<p>Phylogenies can be computed for pathogen genomes that emerge throughout the
simulation. See <code class="docutils literal notranslate"><span class="pre">fitness_function_mutation_example.py</span></code> in the
<code class="docutils literal notranslate"><span class="pre">examples/tutorials/evolution</span></code> folder.
<img alt="Compartments" src="_images/fitness_function_mutation_example_clustermap.png" /></p>
<p>For advanced examples (including multiple parameter sweeps), check out
<a class="reference external" href="https://github.com/pablocarderam/fitness-valleys-opqua">this separate repository</a>
(preprint forthcoming).</p>
</section>
</section>
<section id="requirements-and-installation">
<h2>Requirements and Installation<a class="headerlink" href="#requirements-and-installation" title="Link to this heading"></a></h2>
<p>Opqua runs on Python. A good place to get the latest version it if you don’t
have it is <a class="reference external" href="https://www.anaconda.com/distribution/">Anaconda</a>.</p>
<p>Opqua is <a class="reference external" href="https://pypi.org/project/opqua/">available on PyPI</a> to install
through <code class="docutils literal notranslate"><span class="pre">pip</span></code>, as explained below.</p>
<p>If you haven’t yet, <a class="reference external" href="https://pip.pypa.io/en/stable/installing/">install pip</a>:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>curl<span class="w"> </span>https://bootstrap.pypa.io/get-pip.py<span class="w"> </span>-o<span class="w"> </span>get-pip.py
python<span class="w"> </span>get-pip.py
</pre></div>
</div>
<p>Install Opqua by running</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>pip<span class="w"> </span>install<span class="w"> </span>opqua
</pre></div>
</div>
<p>The pip installer should take care of installing the necessary packages.
However, for reference, the versions of the packages used for opqua’s
development are saved in <code class="docutils literal notranslate"><span class="pre">requirements.txt</span></code></p>
</section>
<section id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Link to this heading"></a></h2>
<p>To run any Opqua model (including the tutorials in the <code class="docutils literal notranslate"><span class="pre">examples/tutorials</span></code>
folder), save the model as a <code class="docutils literal notranslate"><span class="pre">.py</span></code> file and execute from the console using
<code class="docutils literal notranslate"><span class="pre">python</span> <span class="pre">my_model.py</span></code>.</p>
<p>You may also run the models from a notebook environment
such as <a class="reference external" href="https://jupyter.org/">Jupyter</a> or an integrated development environment
(IDE) such as <a class="reference external" href="https://www.spyder-ide.org/">Spyder</a>, both available through
<a class="reference external" href="https://www.anaconda.com/distribution/">Anaconda</a>.</p>
<section id="minimal-example">
<h3>Minimal example<a class="headerlink" href="#minimal-example" title="Link to this heading"></a></h3>
<p>The simplest model you can make using Opqua looks like this:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># This simulates a pathogen with genome &quot;AAAAAAAAAA&quot; spreading in a single</span>
<span class="c1"># population of 100 hosts, 20 of which are initially infected, under example</span>
<span class="c1"># preset conditions for host-host transmission.</span>

<span class="kn">from</span> <span class="nn">opqua.model</span> <span class="kn">import</span> <span class="n">Model</span>

<span class="n">my_model</span> <span class="o">=</span> <span class="n">Model</span><span class="p">()</span>
<span class="n">my_model</span><span class="o">.</span><span class="n">newSetup</span><span class="p">(</span><span class="s1">&#39;my_setup&#39;</span><span class="p">,</span> <span class="n">preset</span><span class="o">=</span><span class="s1">&#39;host-host&#39;</span><span class="p">)</span>
<span class="n">my_model</span><span class="o">.</span><span class="n">newPopulation</span><span class="p">(</span><span class="s1">&#39;my_population&#39;</span><span class="p">,</span> <span class="s1">&#39;my_setup&#39;</span><span class="p">,</span> <span class="n">num_hosts</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">my_model</span><span class="o">.</span><span class="n">addPathogensToHosts</span><span class="p">(</span> <span class="s1">&#39;my_population&#39;</span><span class="p">,{</span><span class="s1">&#39;AAAAAAAAAA&#39;</span><span class="p">:</span><span class="mi">20</span><span class="p">}</span> <span class="p">)</span>
<span class="n">my_model</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">100</span><span class="p">)</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">my_model</span><span class="o">.</span><span class="n">saveToDataFrame</span><span class="p">(</span><span class="s1">&#39;my_model.csv&#39;</span><span class="p">)</span>
<span class="n">graph</span> <span class="o">=</span> <span class="n">my_model</span><span class="o">.</span><span class="n">compartmentPlot</span><span class="p">(</span><span class="s1">&#39;my_model.png&#39;</span><span class="p">,</span> <span class="n">data</span><span class="p">)</span>
</pre></div>
</div>
<p>For more example usage, have a look at the <code class="docutils literal notranslate"><span class="pre">examples</span></code> folder. For an overview
of how Opqua models work, check out the Materials and Methods section on the
manuscript
<a class="reference external" href="https://www.biorxiv.org/content/10.1101/2021.12.16.473045v1.full">here</a>. A
summarized description is shown below in the
<a class="reference internal" href="#how-does-opqua-work"><span class="std std-ref">How Does Opqua Work?</span></a> section.
For more information on the details of each function, head over to the
Model Documentation section.</p>
</section>
</section>
<section id="how-does-opqua-work">
<h2>How Does Opqua Work?<a class="headerlink" href="#how-does-opqua-work" title="Link to this heading"></a></h2>
<section id="basic-concepts">
<h3>Basic concepts<a class="headerlink" href="#basic-concepts" title="Link to this heading"></a></h3>
<p>Opqua models are composed of populations containing hosts and/or vectors, which
themselves may be infected by a number of pathogens with different genomes.</p>
<p>A genome is represented as a string of characters. All genomes must be of the
same length (a set number of loci), and each position within the genome can have
one of a number of different characters specified by the user (corresponding to
different alleles). Different loci in the genome may have different possible
alleles available to them. Genomes may be composed of separate chromosomes,
separated by the “/” character, which is reserved for this purpose.</p>
<p>Each population may have its own unique parameters dictating the events that
happen inside of it, including how pathogens are spread between its hosts and
vectors.</p>
</section>
<section id="events">
<h3>Events<a class="headerlink" href="#events" title="Link to this heading"></a></h3>
<p>There are different kinds of events that may occur to hosts and vectors in
a population:</p>
<ul class="simple">
<li><p>contact between an infectious host/vector and another host/vector in the same
population (intra-population contact) or in a different population (“population
contact”)</p></li>
<li><p>migration of a host/vector from one population to another</p></li>
<li><p>recovery of an infected host/vector</p></li>
<li><p>birth of a new host/vector from an existing host/vector</p></li>
<li><p>death of a host/vector due to pathogen infection or by “natural” causes</p></li>
<li><p>mutation of a pathogen in an infected host/vector</p></li>
<li><p>recombination of two pathogens in an infected host/vector</p></li>
</ul>
<p><img alt="Events" src="img/events.png" /></p>
<p>The likelihood of each event occurring is determined by the population’s
parameters (explained in the <a class="reference internal" href="#newsetup"><span class="std std-ref">newSetup</span></a> function documentation) and
the number of infected and healthy hosts and/or vectors in the population(s)
involved. Crucially, it is also determined by the genome sequences of the
pathogens infecting those hosts and vectors. The user may specify arbitrary
functions to evaluate how a genome sequence affects any of the above kinds of
rates. This is once again done through arguments of the <a class="reference internal" href="#newsetup"><span class="std std-ref">newSetup</span></a>
function. As an example, a specific genome sequence may result in increased
transmission within populations but decreased migration of infected hosts, or
increased mutation rates. These custom functions may be different across
populations, resulting in different adaptive landscapes within different
populations.</p>
<p>Contacts within and between populations may happen by any combination of
host-host, host-vector, and/or vector-host routes, depending on the populations’
parameters. When a contact occurs, each pathogen genome present in the infecting
host/vector may be transferred to the receiving host/vector as long as one
“infectious unit” is inoculated. The number of infectious units inoculated is
randomly distributed based on a Poisson probability distribution. The mean of
this distribution is set by the receiving host/vector’s population parameters,
and is multiplied by the fraction of total intra-host fitness of each pathogen
genome. For instance, consider the mean inoculum size for a host in a given
population is 10 units and the infecting host/vector has two pathogens with
fitnesses of 0.3 and 0.7, respectively. This would make the means of the Poisson
distributions used to generate random infections for each pathogen equal to 3
and 7, respectively.</p>
<p>Inter-population contacts occur via the same mechanism as intra-population
contacts, with the distinction that the two populations must be linked in a
compatible way. As an example, if a vector-borne model with two separate
populations is to allow vectors from Population A to contact hosts in Population
B, then the contact rate of vectors in Population A and the contact rate of
hosts in Population B must both be greater than zero. Migration of hosts/vectors
from one population to another depends on a single rate defining the frequency
of vector/host transport events from a given population to another. Therefore,
Population A would have a specific migration rate dictating transport to
Population B, and Population B would have a separate rate governing transport
towards A.</p>
<p>Recovery of an infected host or vector results in all pathogens being removed
from the host/vector. Additionally, the host/vector may optionally gain
protection from pathogens that contain specific genome sequences present in the
genomes of the pathogens it recovered from, representing immune memory. The user
may specify a population parameter delimiting the contiguous loci in the genome
that are saved on the recovered host/vector as “protection sequences”. Pathogens
containing any of the host/vector’s protection sequences will not be able to
infect the host/vector.</p>
<p>Births result in a new host/vector that may optionally inherit its parent’s
protection sequences. Additionally, a parent may optionally infect its offspring
at birth following a Poisson sampling process equivalent to the one described
for other contact events above. Deaths of existing hosts/vectors can occur both
naturally or due to infection mortality. Only deaths due to infection are
tracked and recorded in the model’s history.</p>
<p>De novo mutation of a pathogen in a given host/vector results in a single locus
within a pathogen’s genome being randomly assigned a new allele from the
possible alleles at that position. Recombination of two pathogens in a given
host/vector creates two new genomes based on the independent segregation of
chromosomes (or reassortment of genome segments, depending on the field) from
the two parent genomes. In addition, there may be a Poisson-distributed random
number of crossover events between homologous parent chromosomes. Recombination
by crossover event will result in all the loci in the chromosome on one side of
the crossover event location being inherited from one of the parents, while the
remainder of the chromosome is inherited from the other parent. The locations of
crossover events are distributed throughout the genome following a uniform
random distribution.</p>
</section>
<section id="interventions">
<h3>Interventions<a class="headerlink" href="#interventions" title="Link to this heading"></a></h3>
<p>Furthermore, the user may specify changes in model behavior at specific
timepoints during the simulation. These changes are known as “interventions”.
Interventions can include any kind of manipulation to populations in the model,
including:</p>
<ul class="simple">
<li><p>adding new populations</p></li>
<li><p>changing a population’s event parameters and adaptive landscape functions</p></li>
<li><p>linking and unlinking populations through migration or inter-population
contact</p></li>
<li><p>adding and removing hosts and vectors to a population</p></li>
</ul>
<p>Interventions can also include actions that involve specific hosts or vectors in
a given population, such as:</p>
<ul class="simple">
<li><p>adding pathogens with specific genomes to a host/vector</p></li>
<li><p>removing all protection sequences from some hosts/vectors in a population</p></li>
<li><p>applying a “treatment” in a population that cures some of its hosts/vectors of
pathogens</p></li>
<li><p>applying a “vaccine” in a population that protects some of its hosts/vectors
from pathogens</p></li>
</ul>
<p>For these kinds of interventions involving specific pathogens in a population,
the user may choose to apply them to a randomly-sampled fraction of
hosts/vectors in a population, or to a specific group of individuals. This is
useful when simulating consecutive interventions on the same specific group
within a population. A single model may contain multiple groups of individuals
and the same individual may be a member of multiple different groups.
Individuals remain in the same group even if they migrate away from the
population they were chosen in.</p>
<p>When a host/vector is given a “treatment”, it removes all pathogens within the
host/vector that do not contain a collection of “resistance sequences”. A
treatment may have multiple resistance sequences. A pathogen must contain all
of these within its genome in order to avoid being removed. On the other hand,
applying a vaccine consists of adding a specific protection sequence to
hosts/vectors, which behaves as explained above for recovered hosts/vectors when
they acquire immune protection, if the model allows it.</p>
</section>
<section id="simulation">
<h3>Simulation<a class="headerlink" href="#simulation" title="Link to this heading"></a></h3>
<p>Models are simulated using an implementation of the Gillespie algorithm in which
the rates of different kinds of events across different populations are
computed with each population’s parameters and current state, and are then
stored in a matrix. In addition, each population has host and vector matrices
containing coefficients that represent the contribution of each host and vector,
respectively, to the rates in the master model rate matrix. Each coefficient is
dependent on the genomes of the pathogens infecting its corresponding vector or
host. Whenever an event occurs, the corresponding entries in the population
matrix are updated, and the master rate matrix is recomputed based on this
information.</p>
<p><img alt="Simulation" src="_images/simulation.png" /></p>
<p>The model’s state at any given time comprises all populations, their hosts
and vectors, and the pathogen genomes infecting each of these. A copy of the
model’s state is saved at every time point, or at intermittent intervals
throughout the course of the simulation. A random sample of hosts and/or vectors
may be saved instead of the entire model as a means of reducing memory
footprint.</p>
</section>
<section id="output">
<h3>Output<a class="headerlink" href="#output" title="Link to this heading"></a></h3>
<p>The output of a model can be saved in multiple ways. The model state at each
saved timepoint may be output in a single, raw <a class="reference internal" href="#pandas.pydata.org/"><span class="xref myst">pandas</span></a>
DataFrame, and saved as a tabular file. Other data output
types include counts of pathogen genomes or protection sequences for the
model, as well as time of first emergence for each pathogen genome and genome
distance matrices for every timepoint sampled. The user can also create
different kinds of plots to visualize the results. These include:</p>
<ul class="simple">
<li><p>plots of the number of hosts and/or vectors in different epidemiological
compartments (naive, infected, recovered, and dead) across simulation time</p></li>
<li><p>plots of the number of individuals in a compartment for different populations</p></li>
<li><p>plots of the genomic composition of the pathogen population over time</p></li>
<li><p>phylogenies of pathogen genomes</p></li>
</ul>
<p>Users can also use the data output formats to make their own custom plots.</p>
</section>
</section>
<section id="model-documentation">
<h2>Model Documentation<a class="headerlink" href="#model-documentation" title="Link to this heading"></a></h2>
<p>All usage is handled through the Opqua <code class="docutils literal notranslate"><span class="pre">Model</span></code> class.
The <code class="docutils literal notranslate"><span class="pre">Model</span></code> class contains populations, setups, and interventions to be used
in simulation. It also contains groups of hosts/vectors for manipulations and
stores model history as snapshots for specific time points.</p>
<p>To use it, import the class as</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">opqua.model</span> <span class="kn">import</span> <span class="n">Model</span>
</pre></div>
</div>
<p>You can find a detailed account of everything <code class="docutils literal notranslate"><span class="pre">Model</span></code> does in the
Model attributes and
<a class="reference internal" href="#model-class-methods-list"><span class="std std-ref">Model class methods list</span></a> sections.</p>
<section id="model-class-attributes">
<h3>Model class attributes<a class="headerlink" href="#model-class-attributes" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p>populations – dictionary with keys=population IDs, values=Population
objects</p></li>
<li><p>setups – dictionary with keys=setup IDs, values=Setup objects</p></li>
<li><p>interventions – contains model interventions in the order they will occur</p></li>
<li><p>groups – dictionary with keys=group IDs, values=lists of hosts/vectors</p></li>
<li><p>history – dictionary with keys=time values, values=Model objects that
are snapshots of Model at that timepoint</p></li>
<li><p>global_trackers – dictionary keeping track of some global indicators over all
the course of the simulation</p></li>
<li><p>custom_condition_trackers – dictionary with keys=ID of custom condition,
values=functions that take a Model object as argument and return True or
False; every time True is returned by a function in
custom_condition_trackers, the simulation time will be stored under the
corresponding ID inside global_trackers[‘custom_condition’]</p></li>
<li><p>t_var – variable that tracks time in simulations</p></li>
</ul>
<p>The dictionary global_trackers contains the following keys:</p>
<ul class="simple">
<li><p>num_events: dictionary with the number of each kind of event in the simulation</p></li>
<li><p>last_event_time: time point at which the last event in the simulation happened</p></li>
<li><p>genomes_seen: list of all unique genomes that have appeared in the
simulation</p></li>
<li><p>custom_conditions: dictionary with keys=ID of custom condition, values=lists
of times; every time True is returned by a function in
custom_condition_trackers, the simulation time will be stored under the
corresponding ID inside global_trackers[‘custom_condition’]</p></li>
</ul>
<p>The dictionary num_events inside of global_trackers contains the following keys:</p>
<ul class="simple">
<li><p>MIGRATE_HOST</p></li>
<li><p>MIGRATE_VECTOR</p></li>
<li><p>POPULATION_CONTACT_HOST_HOST</p></li>
<li><p>POPULATION_CONTACT_HOST_VECTOR</p></li>
<li><p>POPULATION_CONTACT_VECTOR_HOST</p></li>
<li><p>CONTACT_HOST_HOST</p></li>
<li><p>CONTACT_HOST_VECTOR</p></li>
<li><p>CONTACT_VECTOR_HOST</p></li>
<li><p>RECOVER_HOST</p></li>
<li><p>RECOVER_VECTOR</p></li>
<li><p>MUTATE_HOST</p></li>
<li><p>MUTATE_VECTOR</p></li>
<li><p>RECOMBINE_HOST</p></li>
<li><p>RECOMBINE_VECTOR</p></li>
<li><p>KILL_HOST</p></li>
<li><p>KILL_VECTOR</p></li>
<li><p>DIE_HOST</p></li>
<li><p>DIE_VECTOR</p></li>
<li><p>BIRTH_HOST</p></li>
<li><p>BIRTH_VECTOR</p></li>
</ul>
<p>KILL_HOST and KILL_VECTOR denote death due to infection, whereas DIE_HOST and
DIE_VECTOR denote death by natural means.</p>
</section>
<section id="model-class-methods-list">
<h3>Model class methods list<a class="headerlink" href="#model-class-methods-list" title="Link to this heading"></a></h3>
<section id="model-initialization-and-simulation">
<h4>Model initialization and simulation<a class="headerlink" href="#model-initialization-and-simulation" title="Link to this heading"></a></h4>
<ul class="simple">
<li><p><a class="reference internal" href="#setrandomseed"><span class="std std-ref">setRandomSeed</span></a> – set random seed for numpy random number
generator</p></li>
<li><p><a class="reference internal" href="#newsetup"><span class="std std-ref">newSetup</span></a> – creates a new Setup, save it in setups dict under
given name</p></li>
<li><p><a class="reference internal" href="#newintervention"><span class="std std-ref">newIntervention</span></a> – creates a new intervention executed
during simulation</p></li>
<li><p><a class="reference internal" href="#run"><span class="std std-ref">run</span></a> – simulates model for a specified length of time</p></li>
<li><p><a class="reference internal" href="#runreplicates"><span class="std std-ref">runReplicates</span></a> – simulate replicates of a model, save only
end results</p></li>
<li><p><a class="reference internal" href="#runparamsweep"><span class="std std-ref">runParamSweep</span></a> – simulate parameter sweep with a model, save
only end results</p></li>
<li><p><a class="reference internal" href="#copystate"><span class="std std-ref">copyState</span></a> – copies a slimmed-down representation of model state</p></li>
<li><p><a class="reference internal" href="#deepcopy"><span class="std std-ref">deepCopy</span></a> – copies current model with inner references</p></li>
</ul>
</section>
<section id="data-output-and-plotting">
<h4>Data Output and Plotting<a class="headerlink" href="#data-output-and-plotting" title="Link to this heading"></a></h4>
<ul class="simple">
<li><p><a class="reference internal" href="#savetodataframe"><span class="std std-ref">saveToDataFrame</span></a> – saves status of model to data frame,
writes to file</p></li>
<li><p><a class="reference internal" href="#getpathogens"><span class="std std-ref">getPathogens</span></a> – creates data frame with counts for all
pathogen genomes</p></li>
<li><p><a class="reference internal" href="#getprotections"><span class="std std-ref">getProtections</span></a> – creates data frame with counts for all
protection sequences</p></li>
<li><p><a class="reference internal" href="#populationsplot"><span class="std std-ref">populationsPlot</span></a> – plots aggregated totals per
population across time</p></li>
<li><p><a class="reference internal" href="#compartmentplot"><span class="std std-ref">compartmentPlot</span></a> – plots number of naive, infected,
recovered, dead hosts/vectors vs time</p></li>
<li><p><a class="reference internal" href="#compositionplot"><span class="std std-ref">compositionPlot</span></a> – plots counts for pathogen genomes or
resistance vs. time</p></li>
<li><p><a class="reference internal" href="#clustermap"><span class="std std-ref">clustermap</span></a> – plots heatmap and dendrogram of all pathogens in
given data</p></li>
<li><p><a class="reference internal" href="#pathogendistancehistory"><span class="std std-ref">pathogenDistanceHistory</span></a> – calculates pairwise
distances for pathogen genomes at different times</p></li>
<li><p><a class="reference internal" href="#getgenometimes"><span class="std std-ref">getGenomeTimes</span></a> – create DataFrame with times genomes first
appeared during simulation</p></li>
<li><p><a class="reference internal" href="#getcompositiondata"><span class="std std-ref">getCompositionData</span></a> – create dataframe with counts for
pathogen genomes or resistance</p></li>
</ul>
</section>
<section id="model-interventions">
<h4>Model interventions<a class="headerlink" href="#model-interventions" title="Link to this heading"></a></h4>
<p>Make and connect populations:</p>
<ul class="simple">
<li><p><a class="reference internal" href="#newpopulation"><span class="std std-ref">newPopulation</span></a> – create a new Population object with
setup parameters</p></li>
<li><p><a class="reference internal" href="#linkpopulationshostmigration"><span class="std std-ref">linkPopulationsHostMigration</span></a> – set host
migration rate from one population towards another</p></li>
<li><p><a class="reference internal" href="#linkpopulationsvectormigration"><span class="std std-ref">linkPopulationsVectorMigration</span></a> – set
vector migration rate from one population towards another</p></li>
<li><p><a class="reference internal" href="#linkpopulationshosthostcontact"><span class="std std-ref">linkPopulationsHostHostContact</span></a> – set
host-host inter-population contact rate from one population towards another</p></li>
<li><p><a class="reference internal" href="#linkpopulationshostvectorcontact"><span class="std std-ref">linkPopulationsHostVectorContact</span></a> – set
host-vector inter-population contact rate from one population towards another</p></li>
<li><p><a class="reference internal" href="#linkpopulationsvectorhostcontact"><span class="std std-ref">linkPopulationsVectorHostContact</span></a> – set
vector-host inter-population contact rate from one population towards another</p></li>
<li><p><a class="reference internal" href="#createinterconnectedpopulations"><span class="std std-ref">createInterconnectedPopulations</span></a> –
create new populations, link all of them to each other by migration and/or
inter-population contact</p></li>
</ul>
<p>Manipulate hosts and vectors in population:</p>
<ul class="simple">
<li><p><a class="reference internal" href="#newhostgroup"><span class="std std-ref">newHostGroup</span></a> – returns a list of random (healthy or any)
hosts</p></li>
<li><p><a class="reference internal" href="#newvectorgroup"><span class="std std-ref">newVectorGroup</span></a> – returns a list of random (healthy or
any) vectors</p></li>
<li><p><a class="reference internal" href="#addhosts"><span class="std std-ref">addHosts</span></a> – adds hosts to the population</p></li>
<li><p><a class="reference internal" href="#addvectors"><span class="std std-ref">addVectors</span></a> – adds vectors to the population</p></li>
<li><p><a class="reference internal" href="#removehosts"><span class="std std-ref">removeHosts</span></a> – removes hosts from the population</p></li>
<li><p><a class="reference internal" href="#removevectors"><span class="std std-ref">removeVectors</span></a> – removes vectors from the population</p></li>
<li><p><a class="reference internal" href="#addpathogenstohosts"><span class="std std-ref">addPathogensToHosts</span></a> – adds pathogens with
specified genomes to hosts</p></li>
<li><p><a class="reference internal" href="#addpathogenstovectors"><span class="std std-ref">addPathogensToVectors</span></a> – adds pathogens with
specified genomes to vectors</p></li>
<li><p><a class="reference internal" href="#treathosts"><span class="std std-ref">treatHosts</span></a> – removes infections susceptible to given
treatment from hosts</p></li>
<li><p><a class="reference internal" href="#treatvectors"><span class="std std-ref">treatVectors</span></a> – removes infections susceptible to
treatment from vectors</p></li>
<li><p><a class="reference internal" href="#protecthosts"><span class="std std-ref">protectHosts</span></a> – adds protection sequence to hosts</p></li>
<li><p><a class="reference internal" href="#protectvectors"><span class="std std-ref">protectVectors</span></a> – adds protection sequence to vectors</p></li>
<li><p><a class="reference internal" href="#wipeprotectionhosts"><span class="std std-ref">wipeProtectionHosts</span></a> – removes all protection
sequences from hosts</p></li>
<li><p><a class="reference internal" href="#wipeprotectionvectors"><span class="std std-ref">wipeProtectionVectors</span></a> – removes all protection
sequences from vectors</p></li>
</ul>
<p>Modify population parameters:</p>
<ul class="simple">
<li><p><a class="reference internal" href="#setsetup"><span class="std std-ref">setSetup</span></a> – assigns a given set of parameters to this
population</p></li>
</ul>
<p>Utility:</p>
<ul class="simple">
<li><p><a class="reference internal" href="#custommodelfunction"><span class="std std-ref">customModelFunction</span></a> – returns output of given
function run on model</p></li>
</ul>
</section>
<section id="preset-fitness-functions">
<h4>Preset fitness functions<a class="headerlink" href="#preset-fitness-functions" title="Link to this heading"></a></h4>
<ul class="simple">
<li><p><a class="reference internal" href="#peaklandscape"><span class="std std-ref">peakLandscape</span></a> – evaluates genome numeric phenotype by
decreasing with distance from optimal sequence</p></li>
<li><p><a class="reference internal" href="#valleylandscape"><span class="std std-ref">valleyLandscape</span></a> – evaluates genome numeric phenotype by
increasing with distance from worst sequence</p></li>
</ul>
</section>
</section>
<section id="detailed-model-method-list">
<h3>Detailed Model method list<a class="headerlink" href="#detailed-model-method-list" title="Link to this heading"></a></h3>
<section id="setrandomseed">
<h4>setRandomSeed<a class="headerlink" href="#setrandomseed" title="Link to this heading"></a></h4>
<dl class="py function">
<dt class="sig sig-object py" id="opqua.model.Model.setRandomSeed">
<span class="sig-prename descclassname"><span class="pre">opqua.model.Model.</span></span><span class="sig-name descname"><span class="pre">setRandomSeed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#opqua.model.Model.setRandomSeed" title="Link to this definition"></a></dt>
<dd><p>Set random seed for numpy random number generator.</p>
<p>Arguments:
seed – int for the random seed to be passed to numpy (int)</p>
</dd></dl>

</section>
<section id="model">
<h4>Model<a class="headerlink" href="#model" title="Link to this heading"></a></h4>
<dl class="py function">
<dt class="sig sig-object py" id="opqua.model.Model">
<span class="sig-prename descclassname"><span class="pre">opqua.model.</span></span><span class="sig-name descname"><span class="pre">Model</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/model.html#Model"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.model.Model" title="Link to this definition"></a></dt>
<dd><p>Class defines a Model.</p>
<p>This is the main class that the user interacts with.</p>
<p>The Model class contains populations, setups, and interventions to be used
in simulation. Also contains groups of hosts/vectors for manipulations and
stores model history as snapshots for each time point.</p>
<p><strong>* — CONSTANTS: — *</strong>
### Color scheme constants ###
CB_PALETTE – a colorblind-friendly 8-color color scheme
DEF_CMAP – a colormap object for Seaborn plots</p>
<p><strong>* — ATTRIBUTES: — *</strong>
populations – dictionary with keys=population IDs, values=Population</p>
<blockquote>
<div><p>objects</p>
</div></blockquote>
<p>setups – dictionary with keys=setup IDs, values=Setup objects
interventions – contains model interventions in the order they will occur
groups – dictionary with keys=group IDs, values=lists of hosts/vectors
history – dictionary with keys=time values, values=Model objects that</p>
<blockquote>
<div><p>are snapshots of Model at that timepoint</p>
</div></blockquote>
<p>t_var – variable that tracks time in simulations</p>
<p><strong>* — METHODS: — *</strong></p>
<p>— Model Initialization and Simulation: —</p>
<p>setRandomSeed – set random seed for numpy random number generator
newSetup – creates a new Setup, save it in setups dict under given name
newIntervention – creates a new intervention executed during simulation
run – simulates model for a specified length of time
runReplicates – simulate replicates of a model, save only end results
runParamSweep – simulate parameter sweep with a model, save only end results
copyState – returns a slimmed-down version of the current model state</p>
<p>— Data Output and Plotting: —</p>
<p>saveToDataFrame – saves status of model to dataframe, writes to file
getPathogens – creates Dataframe with counts for all pathogen genomes
getProtections – creates Dataframe with counts for all protection sequences
populationsPlot – plots aggregated totals per population across time
compartmentPlot – plots number of naive,inf,rec,dead hosts/vectors vs time
compositionPlot – plots counts for pathogen genomes or resistance vs. time
clustermap – create a heatmap and dendrogram for pathogen genomes in data
pathogenDistanceHistory – get pairwise distances for pathogen genomes
getGenomeTimes – create DataFrame with times genomes first appeared during</p>
<blockquote>
<div><p>simulation</p>
</div></blockquote>
<dl class="simple">
<dt>getCompositionData – create dataframe with counts for pathogen genomes or</dt><dd><p>resistance</p>
</dd>
</dl>
<p>— Model interventions: —</p>
<ul class="simple">
<li><p>Make and connect populations -</p></li>
</ul>
<p>newPopulation – create a new Population object with setup parameters
linkPopulationsHostMigration – set host migration rate from one population</p>
<blockquote>
<div><p>towards another</p>
</div></blockquote>
<dl class="simple">
<dt>linkPopulationsVectorMigration – set vector migration rate from one</dt><dd><p>population towards another</p>
</dd>
<dt>linkPopulationsHostHostContact – set host-host inter-population contact</dt><dd><p>rate from one population towards another</p>
</dd>
<dt>linkPopulationsHostVectorMigration – set host-vector inter-population</dt><dd><p>contact rate from one population towards another</p>
</dd>
<dt>linkPopulationsVectorHostMigration – set vector-host inter-population</dt><dd><p>contact rate from one population towards another</p>
</dd>
<dt>createInterconnectedPopulations – create new populations, link all of them</dt><dd><p>to each other by migration and/or inter-population contact</p>
</dd>
</dl>
<ul class="simple">
<li><p>Manipulate hosts and vectors in population -</p></li>
</ul>
<p>newHostGroup – returns a list of random (healthy or any) hosts
newVectorGroup – returns a list of random (healthy or any) vectors
addHosts – adds hosts to the population
addVectors – adds vectors to the population
removeHosts – removes hosts from the population
removeVectors – removes vectors from the population
addPathogensToHosts – adds pathogens with specified genomes to hosts
addPathogensToVectors – adds pathogens with specified genomes to vectors
treatHosts – removes infections susceptible to given treatment from hosts
treatVectors – removes infections susceptible to treatment from vectors
protectHosts – adds protection sequence to hosts
protectVectors – adds protection sequence to vectors
wipeProtectionHosts – removes all protection sequences from hosts
wipeProtectionVectors – removes all protection sequences from vectors</p>
<ul class="simple">
<li><p>Modify population parameters -</p></li>
</ul>
<p>setSetup – assigns a given set of parameters to this population</p>
<ul class="simple">
<li><p>Utility -</p></li>
</ul>
<p>customModelFunction – returns output of given function run on model</p>
<dl class="simple">
<dt>— Preset fitness functions: —</dt><dd><ul class="simple">
<li><p>these are static methods</p></li>
</ul>
</dd>
<dt>peakLandscape – evaluates genome numeric phenotype by decreasing with</dt><dd><p>distance from optimal sequence</p>
</dd>
<dt>valleyLandscape – evaluates genome numeric phenotype by increasing with</dt><dd><p>distance from worst sequence</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="newsetup">
<h4>newSetup<a class="headerlink" href="#newsetup" title="Link to this heading"></a></h4>
<dl class="py function">
<dt class="sig sig-object py" id="opqua.model.Model.newSetup">
<span class="sig-prename descclassname"><span class="pre">opqua.model.Model.</span></span><span class="sig-name descname"><span class="pre">newSetup</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">preset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_loci</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">possible_alleles</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fitnessHost</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">contactHost</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">receiveContactHost</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mortalityHost</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">natalityHost</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recoveryHost</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">migrationHost</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">populationContactHost</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">receivePopulationContactHost</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutationHost</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recombinationHost</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fitnessVector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">contactVector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">receiveContactVector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mortalityVector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">natalityVector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recoveryVector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">migrationVector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">populationContactVector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">receivePopulationContactVector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutationVector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recombinationVector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">contact_rate_host_vector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transmission_efficiency_host_vector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transmission_efficiency_vector_host</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">contact_rate_host_host</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transmission_efficiency_host_host</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mean_inoculum_host</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mean_inoculum_vector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recovery_rate_host</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recovery_rate_vector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mortality_rate_host</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mortality_rate_vector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recombine_in_host</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recombine_in_vector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_crossover_host</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_crossover_vector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutate_in_host</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutate_in_vector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">death_rate_host</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">death_rate_vector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">birth_rate_host</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">birth_rate_vector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vertical_transmission_host</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vertical_transmission_vector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inherit_protection_host</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inherit_protection_vector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">protection_upon_recovery_host</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">protection_upon_recovery_vector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#opqua.model.Model.newSetup" title="Link to this definition"></a></dt>
<dd><p>Create a new Setup, save it in setups dict under given name.</p>
<p>Two preset setups exist: “vector-borne” and “host-host”. You may select
one of the preset setups with the preset keyword argument and then
modify individual parameters with additional keyword arguments, without
having to specify all of them.</p>
<p>Arguments:
name – name of setup to be used as a key in model setups dictionary</p>
<p>Keyword arguments:
preset – preset setup to be used: “vector-borne” or “host-host”, if</p>
<blockquote>
<div><p>None, must define all other keyword arguments (default None; None or
String)</p>
</div></blockquote>
<p>num_loci – length of each pathogen genome string (int &gt; 0)
possible_alleles – set of possible characters in all genome string, or</p>
<blockquote>
<div><p>at each position in genome string (String or list of Strings with
num_loci elements)</p>
</div></blockquote>
<dl class="simple">
<dt>fitnessHost – function that evaluates relative fitness in head-to-head</dt><dd><p>competition for different genomes within the same host
(function object, takes a String argument and returns a number &gt;= 0)</p>
</dd>
<dt>contactHost – function that returns coefficient modifying probability</dt><dd><p>of a given host being chosen to be the infector in a contact event,
based on genome sequence of pathogen
(function object, takes a String argument and returns a number 0-1)</p>
</dd>
<dt>receiveContactHost – function that returns coefficient modifying</dt><dd><p>probability of a given host being chosen to be the infected in
a contact event, based on genome sequence of pathogen</p>
</dd>
<dt>mortalityHost – function that returns coefficient modifying death rate</dt><dd><p>for a given host, based on genome sequence of pathogen
(function object, takes a String argument and returns a number 0-1)</p>
</dd>
<dt>natalityHost – function that returns coefficient modifying birth rate</dt><dd><p>for a given host, based on genome sequence of pathogen
(function object, takes a String argument and returns a number 0-1)</p>
</dd>
<dt>recoveryHost – function that returns coefficient modifying recovery</dt><dd><p>rate for a given host based on genome sequence of pathogen
(function object, takes a String argument and returns a number 0-1)</p>
</dd>
<dt>migrationHost – function that returns coefficient modifying migration</dt><dd><p>rate for a given host based on genome sequence of pathogen
(function object, takes a String argument and returns a number 0-1)</p>
</dd>
<dt>populationContactHost – function that returns coefficient modifying</dt><dd><p>population contact rate for a given host based on genome sequence of
pathogen
(function object, takes a String argument and returns a number 0-1)</p>
</dd>
<dt>mutationHost – function that returns coefficient modifying mutation</dt><dd><p>rate for a given host based on genome sequence of pathogen
(function object, takes a String argument and returns a number 0-1)</p>
</dd>
<dt>recombinationHost – function that returns coefficient modifying</dt><dd><p>recombination rate for a given host based on genome sequence of
pathogen
(function object, takes a String argument and returns a number 0-1)</p>
</dd>
<dt>fitnessVector – function that evaluates relative fitness in head-to-</dt><dd><p>head competition for different genomes within the same vector
(function object, takes a String argument and returns a number &gt;= 0)</p>
</dd>
<dt>contactVector – function that returns coefficient modifying probability</dt><dd><p>of a given vector being chosen to be the infector in a contact
event, based on genome sequence of pathogen
(function object, takes a String argument and returns a number 0-1)</p>
</dd>
<dt>receiveContactVector – function that returns coefficient modifying</dt><dd><p>probability of a given vector being chosen to be the infected in
a contact event, based on genome sequence of pathogen
(function object, takes a String argument and returns a number 0-1)</p>
</dd>
<dt>mortalityVector – function that returns coefficient modifying death</dt><dd><p>rate for a given vector, based on genome sequence of pathogen
(function object, takes a String argument and returns a number 0-1)</p>
</dd>
<dt>natalityVector – function that returns coefficient modifying birth rate</dt><dd><p>for a given vector, based on genome sequence of pathogen
(function object, takes a String argument and returns a number 0-1)</p>
</dd>
<dt>recoveryVector – function that returns coefficient modifying recovery</dt><dd><p>rate for a given vector based on genome sequence of pathogen
(function object, takes a String argument and returns a number 0-1)</p>
</dd>
<dt>migrationVector – function that returns coefficient modifying migration</dt><dd><p>rate for a given vector based on genome sequence of pathogen
(function object, takes a String argument and returns a number 0-1)</p>
</dd>
<dt>populationContactVector – function that returns coefficient modifying</dt><dd><p>population contact rate for a given vector based on genome sequence
of pathogen
(function object, takes a String argument and returns a number 0-1)</p>
</dd>
<dt>mutationVector – function that returns coefficient modifying mutation</dt><dd><p>rate for a given vector based on genome sequence of pathogen
(function object, takes a String argument and returns a number 0-1)</p>
</dd>
<dt>recombinationVector – function that returns coefficient modifying</dt><dd><p>recombination rate for a given vector based on genome sequence of
pathogen
(function object, takes a String argument and returns a number 0-1)</p>
</dd>
<dt>contact_rate_host_vector – rate of host-vector contact events, not</dt><dd><p>necessarily transmission, assumes constant population density;
evts/time (number &gt;= 0)</p>
</dd>
<dt>transmission_efficiency_host_vector – fraction of host-vector contacts</dt><dd><p>that result in successful transmission</p>
</dd>
<dt>transmission_efficiency_vector_host – fraction of vector-host contacts</dt><dd><p>that result in successful transmission</p>
</dd>
<dt>contact_rate_host_host – rate of host-host contact events, not</dt><dd><p>necessarily transmission, assumes constant population density;
evts/time (number &gt;= 0)</p>
</dd>
<dt>transmission_efficiency_host_host – fraction of host-host contacts</dt><dd><p>that result in successful transmission</p>
</dd>
<dt>mean_inoculum_host – mean number of pathogens that are transmitted from</dt><dd><p>a vector or host into a new host during a contact event (int &gt;= 0)</p>
</dd>
<dt>mean_inoculum_vector – mean number of pathogens that are transmitted</dt><dd><p>from a host to a vector during a contact event (int &gt;= 0)</p>
</dd>
<dt>recovery_rate_host – rate at which hosts clear all pathogens;</dt><dd><p>1/time (number &gt;= 0)</p>
</dd>
<dt>recovery_rate_vector – rate at which vectors clear all pathogens</dt><dd><p>1/time (number &gt;= 0)</p>
</dd>
<dt>recovery_rate_vector – rate at which vectors clear all pathogens</dt><dd><p>1/time (number &gt;= 0)</p>
</dd>
<dt>mortality_rate_host – rate at which infected hosts die from disease</dt><dd><p>(number 0-1)</p>
</dd>
<dt>mortality_rate_vector – rate at which infected vectors die from</dt><dd><p>disease (number 0-1)</p>
</dd>
<dt>recombine_in_host – rate at which recombination occurs in host;</dt><dd><p>evts/time (number &gt;= 0)</p>
</dd>
<dt>recombine_in_vector – rate at which recombination occurs in vector;</dt><dd><p>evts/time (number &gt;= 0)</p>
</dd>
<dt>num_crossover_host – mean of a Poisson distribution modeling the number</dt><dd><p>of crossover events of host recombination events (number &gt;= 0)</p>
</dd>
<dt>num_crossover_vector – mean of a Poisson distribution modeling the</dt><dd><p>number of crossover events of vector recombination events
(number &gt;= 0)</p>
</dd>
<dt>mutate_in_host – rate at which mutation occurs in host; evts/time</dt><dd><p>(number &gt;= 0)</p>
</dd>
<dt>mutate_in_vector – rate at which mutation occurs in vector; evts/time</dt><dd><p>(number &gt;= 0)</p>
</dd>
</dl>
<p>death_rate_host – natural host death rate; 1/time (number &gt;= 0)
death_rate_vector – natural vector death rate; 1/time (number &gt;= 0)
birth_rate_host – infected host birth rate; 1/time (number &gt;= 0)
birth_rate_vector – infected vector birth rate; 1/time (number &gt;= 0)
vertical_transmission_host – probability that a host is infected by its</p>
<blockquote>
<div><p>parent at birth (number 0-1)</p>
</div></blockquote>
<dl class="simple">
<dt>vertical_transmission_vector – probability that a vector is infected by</dt><dd><p>its parent at birth (number 0-1)</p>
</dd>
<dt>inherit_protection_host – probability that a host inherits all</dt><dd><p>protection sequences from its parent (number 0-1)</p>
</dd>
<dt>inherit_protection_vector – probability that a vector inherits all</dt><dd><p>protection sequences from its parent (number 0-1)</p>
</dd>
<dt>protection_upon_recovery_host – defines indexes in genome string that</dt><dd><p>define substring to be added to host protection sequences after
recovery (None or array-like of length 2 with int 0-num_loci)</p>
</dd>
<dt>protection_upon_recovery_vector – defines indexes in genome string that</dt><dd><p>define substring to be added to vector protection sequences after
recovery (None or array-like of length 2 with int 0-num_loci)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="opqua.model.Model.newSetup">
<span class="sig-prename descclassname"><span class="pre">opqua.model.Model.</span></span><span class="sig-name descname"><span class="pre">newSetup</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">preset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_loci</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">possible_alleles</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fitnessHost</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">contactHost</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">receiveContactHost</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mortalityHost</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">natalityHost</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recoveryHost</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">migrationHost</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">populationContactHost</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">receivePopulationContactHost</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutationHost</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recombinationHost</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fitnessVector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">contactVector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">receiveContactVector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mortalityVector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">natalityVector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recoveryVector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">migrationVector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">populationContactVector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">receivePopulationContactVector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutationVector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recombinationVector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">contact_rate_host_vector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transmission_efficiency_host_vector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transmission_efficiency_vector_host</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">contact_rate_host_host</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transmission_efficiency_host_host</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mean_inoculum_host</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mean_inoculum_vector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recovery_rate_host</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recovery_rate_vector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mortality_rate_host</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mortality_rate_vector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recombine_in_host</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recombine_in_vector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_crossover_host</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_crossover_vector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutate_in_host</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutate_in_vector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">death_rate_host</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">death_rate_vector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">birth_rate_host</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">birth_rate_vector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vertical_transmission_host</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vertical_transmission_vector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inherit_protection_host</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inherit_protection_vector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">protection_upon_recovery_host</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">protection_upon_recovery_vector</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#opqua.model.Model.newSetup" title="Link to this definition"></a></dt>
<dd><p>Create a new Setup, save it in setups dict under given name.</p>
<p>Two preset setups exist: “vector-borne” and “host-host”. You may select
one of the preset setups with the preset keyword argument and then
modify individual parameters with additional keyword arguments, without
having to specify all of them.</p>
<p>Arguments:
name – name of setup to be used as a key in model setups dictionary</p>
<p>Keyword arguments:
preset – preset setup to be used: “vector-borne” or “host-host”, if</p>
<blockquote>
<div><p>None, must define all other keyword arguments (default None; None or
String)</p>
</div></blockquote>
<p>num_loci – length of each pathogen genome string (int &gt; 0)
possible_alleles – set of possible characters in all genome string, or</p>
<blockquote>
<div><p>at each position in genome string (String or list of Strings with
num_loci elements)</p>
</div></blockquote>
<dl class="simple">
<dt>fitnessHost – function that evaluates relative fitness in head-to-head</dt><dd><p>competition for different genomes within the same host
(function object, takes a String argument and returns a number &gt;= 0)</p>
</dd>
<dt>contactHost – function that returns coefficient modifying probability</dt><dd><p>of a given host being chosen to be the infector in a contact event,
based on genome sequence of pathogen
(function object, takes a String argument and returns a number 0-1)</p>
</dd>
<dt>receiveContactHost – function that returns coefficient modifying</dt><dd><p>probability of a given host being chosen to be the infected in
a contact event, based on genome sequence of pathogen</p>
</dd>
<dt>mortalityHost – function that returns coefficient modifying death rate</dt><dd><p>for a given host, based on genome sequence of pathogen
(function object, takes a String argument and returns a number 0-1)</p>
</dd>
<dt>natalityHost – function that returns coefficient modifying birth rate</dt><dd><p>for a given host, based on genome sequence of pathogen
(function object, takes a String argument and returns a number 0-1)</p>
</dd>
<dt>recoveryHost – function that returns coefficient modifying recovery</dt><dd><p>rate for a given host based on genome sequence of pathogen
(function object, takes a String argument and returns a number 0-1)</p>
</dd>
<dt>migrationHost – function that returns coefficient modifying migration</dt><dd><p>rate for a given host based on genome sequence of pathogen
(function object, takes a String argument and returns a number 0-1)</p>
</dd>
<dt>populationContactHost – function that returns coefficient modifying</dt><dd><p>population contact rate for a given host based on genome sequence of
pathogen
(function object, takes a String argument and returns a number 0-1)</p>
</dd>
<dt>mutationHost – function that returns coefficient modifying mutation</dt><dd><p>rate for a given host based on genome sequence of pathogen
(function object, takes a String argument and returns a number 0-1)</p>
</dd>
<dt>recombinationHost – function that returns coefficient modifying</dt><dd><p>recombination rate for a given host based on genome sequence of
pathogen
(function object, takes a String argument and returns a number 0-1)</p>
</dd>
<dt>fitnessVector – function that evaluates relative fitness in head-to-</dt><dd><p>head competition for different genomes within the same vector
(function object, takes a String argument and returns a number &gt;= 0)</p>
</dd>
<dt>contactVector – function that returns coefficient modifying probability</dt><dd><p>of a given vector being chosen to be the infector in a contact
event, based on genome sequence of pathogen
(function object, takes a String argument and returns a number 0-1)</p>
</dd>
<dt>receiveContactVector – function that returns coefficient modifying</dt><dd><p>probability of a given vector being chosen to be the infected in
a contact event, based on genome sequence of pathogen
(function object, takes a String argument and returns a number 0-1)</p>
</dd>
<dt>mortalityVector – function that returns coefficient modifying death</dt><dd><p>rate for a given vector, based on genome sequence of pathogen
(function object, takes a String argument and returns a number 0-1)</p>
</dd>
<dt>natalityVector – function that returns coefficient modifying birth rate</dt><dd><p>for a given vector, based on genome sequence of pathogen
(function object, takes a String argument and returns a number 0-1)</p>
</dd>
<dt>recoveryVector – function that returns coefficient modifying recovery</dt><dd><p>rate for a given vector based on genome sequence of pathogen
(function object, takes a String argument and returns a number 0-1)</p>
</dd>
<dt>migrationVector – function that returns coefficient modifying migration</dt><dd><p>rate for a given vector based on genome sequence of pathogen
(function object, takes a String argument and returns a number 0-1)</p>
</dd>
<dt>populationContactVector – function that returns coefficient modifying</dt><dd><p>population contact rate for a given vector based on genome sequence
of pathogen
(function object, takes a String argument and returns a number 0-1)</p>
</dd>
<dt>mutationVector – function that returns coefficient modifying mutation</dt><dd><p>rate for a given vector based on genome sequence of pathogen
(function object, takes a String argument and returns a number 0-1)</p>
</dd>
<dt>recombinationVector – function that returns coefficient modifying</dt><dd><p>recombination rate for a given vector based on genome sequence of
pathogen
(function object, takes a String argument and returns a number 0-1)</p>
</dd>
<dt>contact_rate_host_vector – rate of host-vector contact events, not</dt><dd><p>necessarily transmission, assumes constant population density;
evts/time (number &gt;= 0)</p>
</dd>
<dt>transmission_efficiency_host_vector – fraction of host-vector contacts</dt><dd><p>that result in successful transmission</p>
</dd>
<dt>transmission_efficiency_vector_host – fraction of vector-host contacts</dt><dd><p>that result in successful transmission</p>
</dd>
<dt>contact_rate_host_host – rate of host-host contact events, not</dt><dd><p>necessarily transmission, assumes constant population density;
evts/time (number &gt;= 0)</p>
</dd>
<dt>transmission_efficiency_host_host – fraction of host-host contacts</dt><dd><p>that result in successful transmission</p>
</dd>
<dt>mean_inoculum_host – mean number of pathogens that are transmitted from</dt><dd><p>a vector or host into a new host during a contact event (int &gt;= 0)</p>
</dd>
<dt>mean_inoculum_vector – mean number of pathogens that are transmitted</dt><dd><p>from a host to a vector during a contact event (int &gt;= 0)</p>
</dd>
<dt>recovery_rate_host – rate at which hosts clear all pathogens;</dt><dd><p>1/time (number &gt;= 0)</p>
</dd>
<dt>recovery_rate_vector – rate at which vectors clear all pathogens</dt><dd><p>1/time (number &gt;= 0)</p>
</dd>
<dt>recovery_rate_vector – rate at which vectors clear all pathogens</dt><dd><p>1/time (number &gt;= 0)</p>
</dd>
<dt>mortality_rate_host – rate at which infected hosts die from disease</dt><dd><p>(number 0-1)</p>
</dd>
<dt>mortality_rate_vector – rate at which infected vectors die from</dt><dd><p>disease (number 0-1)</p>
</dd>
<dt>recombine_in_host – rate at which recombination occurs in host;</dt><dd><p>evts/time (number &gt;= 0)</p>
</dd>
<dt>recombine_in_vector – rate at which recombination occurs in vector;</dt><dd><p>evts/time (number &gt;= 0)</p>
</dd>
<dt>num_crossover_host – mean of a Poisson distribution modeling the number</dt><dd><p>of crossover events of host recombination events (number &gt;= 0)</p>
</dd>
<dt>num_crossover_vector – mean of a Poisson distribution modeling the</dt><dd><p>number of crossover events of vector recombination events
(number &gt;= 0)</p>
</dd>
<dt>mutate_in_host – rate at which mutation occurs in host; evts/time</dt><dd><p>(number &gt;= 0)</p>
</dd>
<dt>mutate_in_vector – rate at which mutation occurs in vector; evts/time</dt><dd><p>(number &gt;= 0)</p>
</dd>
</dl>
<p>death_rate_host – natural host death rate; 1/time (number &gt;= 0)
death_rate_vector – natural vector death rate; 1/time (number &gt;= 0)
birth_rate_host – infected host birth rate; 1/time (number &gt;= 0)
birth_rate_vector – infected vector birth rate; 1/time (number &gt;= 0)
vertical_transmission_host – probability that a host is infected by its</p>
<blockquote>
<div><p>parent at birth (number 0-1)</p>
</div></blockquote>
<dl class="simple">
<dt>vertical_transmission_vector – probability that a vector is infected by</dt><dd><p>its parent at birth (number 0-1)</p>
</dd>
<dt>inherit_protection_host – probability that a host inherits all</dt><dd><p>protection sequences from its parent (number 0-1)</p>
</dd>
<dt>inherit_protection_vector – probability that a vector inherits all</dt><dd><p>protection sequences from its parent (number 0-1)</p>
</dd>
<dt>protection_upon_recovery_host – defines indexes in genome string that</dt><dd><p>define substring to be added to host protection sequences after
recovery (None or array-like of length 2 with int 0-num_loci)</p>
</dd>
<dt>protection_upon_recovery_vector – defines indexes in genome string that</dt><dd><p>define substring to be added to vector protection sequences after
recovery (None or array-like of length 2 with int 0-num_loci)</p>
</dd>
</dl>
</dd></dl>

<p>Two preset setups exist: “vector-borne” and “host-host”. You may select
one of the preset setups with the preset keyword argument and then
modify individual parameters with additional keyword arguments, without
having to specify all of them.</p>
<p><strong>“host-host”</strong>:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">num_loci</span> <span class="o">=</span> <span class="mi">10</span>
<span class="n">possible_alleles</span> <span class="o">=</span> <span class="s1">&#39;ATCG&#39;</span>
<span class="n">fitnessHost</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">contactHost</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">receiveContactHost</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">mortalityHost</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">natalityHost</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">recoveryHost</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">migrationHost</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">populationContactHost</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">receivePopulationContactHost</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">mutationHost</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">recombinationHost</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">fitnessVector</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">contactVector</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">receiveContactVector</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">mortalityVector</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">natalityVector</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">recoveryVector</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">migrationVector</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">populationContactVector</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">receivePopulationContactVector</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">mutationVector</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">recombinationVector</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">contact_rate_host_vector</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">transmission_efficiency_host_vector</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">transmission_efficiency_vector_host</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">contact_rate_host_host</span> <span class="o">=</span> <span class="mf">2e-1</span>
<span class="n">transmission_efficiency_host_host</span> <span class="o">=</span> <span class="mi">1</span>
<span class="n">mean_inoculum_host</span> <span class="o">=</span> <span class="mf">1e1</span>
<span class="n">mean_inoculum_vector</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">recovery_rate_host</span> <span class="o">=</span> <span class="mf">1e-1</span>
<span class="n">recovery_rate_vector</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">mortality_rate_host</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">mortality_rate_vector</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">recombine_in_host</span> <span class="o">=</span> <span class="mf">1e-4</span>
<span class="n">recombine_in_vector</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">num_crossover_host</span> <span class="o">=</span> <span class="mi">1</span>
<span class="n">num_crossover_vector</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">mutate_in_host</span> <span class="o">=</span> <span class="mf">1e-6</span>
<span class="n">mutate_in_vector</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">death_rate_host</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">death_rate_vector</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">birth_rate_host</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">birth_rate_vector</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">vertical_transmission_host</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">vertical_transmission_vector</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">inherit_protection_host</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">inherit_protection_vector</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">protection_upon_recovery_host</span> <span class="o">=</span> <span class="kc">None</span>
<span class="n">protection_upon_recovery_vector</span> <span class="o">=</span> <span class="kc">None</span>
</pre></div>
</div>
<p><strong>“vector-borne”</strong>:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">num_loci</span> <span class="o">=</span> <span class="mi">10</span>
<span class="n">possible_alleles</span> <span class="o">=</span> <span class="s1">&#39;ATCG&#39;</span>
<span class="n">fitnessHost</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">contactHost</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">receiveContactHost</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">mortalityHost</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">natalityHost</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">recoveryHost</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">migrationHost</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">populationContactHost</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">receivePopulationContactHost</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">mutationHost</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">recombinationHost</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">fitnessVector</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">contactVector</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">receiveContactVector</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">mortalityVector</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">natalityVector</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">recoveryVector</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">migrationVector</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">populationContactVector</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">receivePopulationContactVector</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">mutationVector</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">recombinationVector</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">contact_rate_host_vector</span> <span class="o">=</span> <span class="mf">2e-1</span>
<span class="n">transmission_efficiency_host_vector</span> <span class="o">=</span> <span class="mi">1</span>
<span class="n">transmission_efficiency_vector_host</span> <span class="o">=</span> <span class="mi">1</span>
<span class="n">contact_rate_host_host</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">transmission_efficiency_host_host</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">mean_inoculum_host</span> <span class="o">=</span> <span class="mf">1e2</span>
<span class="n">mean_inoculum_vector</span> <span class="o">=</span> <span class="mf">1e0</span>
<span class="n">recovery_rate_host</span> <span class="o">=</span> <span class="mf">1e-1</span>
<span class="n">recovery_rate_vector</span> <span class="o">=</span> <span class="mf">1e-1</span>
<span class="n">mortality_rate_host</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">mortality_rate_vector</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">recombine_in_host</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">recombine_in_vector</span> <span class="o">=</span> <span class="mf">1e-4</span>
<span class="n">num_crossover_host</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">num_crossover_vector</span> <span class="o">=</span> <span class="mi">1</span>
<span class="n">mutate_in_host</span> <span class="o">=</span> <span class="mf">1e-6</span>
<span class="n">mutate_in_vector</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">death_rate_host</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">death_rate_vector</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">birth_rate_host</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">birth_rate_vector</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">vertical_transmission_host</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">vertical_transmission_vector</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">inherit_protection_host</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">inherit_protection_vector</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">protection_upon_recovery_host</span> <span class="o">=</span> <span class="kc">None</span>
<span class="n">protection_upon_recovery_vector</span> <span class="o">=</span> <span class="kc">None</span>
</pre></div>
</div>
</section>
<section id="newintervention">
<h4>newIntervention<a class="headerlink" href="#newintervention" title="Link to this heading"></a></h4>
<dl class="py function">
<dt class="sig sig-object py" id="opqua.model.Model.newIntervention">
<span class="sig-prename descclassname"><span class="pre">opqua.model.Model.</span></span><span class="sig-name descname"><span class="pre">newIntervention</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#opqua.model.Model.newIntervention" title="Link to this definition"></a></dt>
<dd><p>Create a new intervention to be carried out at a specific time.</p>
<p>Arguments:
time – time at which intervention will take place (number &gt;= 0)
method_name – intervention to be carried out, must correspond to the</p>
<blockquote>
<div><p>name of a method of the Model object (String)</p>
</div></blockquote>
<p>args – contains arguments for function in positinal order (array-like)</p>
</dd></dl>

</section>
<section id="addcustomconditiontracker">
<h4>addCustomConditionTracker<a class="headerlink" href="#addcustomconditiontracker" title="Link to this heading"></a></h4>
<dl class="py function">
<dt class="sig sig-object py" id="opqua.model.Model.addCustomConditionTracker">
<span class="sig-prename descclassname"><span class="pre">opqua.model.Model.</span></span><span class="sig-name descname"><span class="pre">addCustomConditionTracker</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">condition_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trackerFunction</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#opqua.model.Model.addCustomConditionTracker" title="Link to this definition"></a></dt>
<dd><p>Add a function to track occurrences of custom events in simulation.</p>
<p>Adds function trackerFunction to dictionary custom_condition_trackers
under key condition_id. Function trackerFunction will be executed at
every event in the simulation. Every time True is returned,
the simulation time will be stored under the corresponding condition_id
key inside global_trackers[‘custom_condition’]</p>
<p>Arguments:
condition_id – ID of this specific condition (String)
trackerFunction – function that take a Model object as argument and</p>
<blockquote>
<div><p>returns True or False; (Function)</p>
</div></blockquote>
</dd></dl>

</section>
<section id="run">
<h4>run<a class="headerlink" href="#run" title="Link to this heading"></a></h4>
<dl class="py function">
<dt class="sig sig-object py" id="opqua.model.Model.run">
<span class="sig-prename descclassname"><span class="pre">opqua.model.Model.</span></span><span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">t0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tf</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time_sampling</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">host_sampling</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vector_sampling</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#opqua.model.Model.run" title="Link to this definition"></a></dt>
<dd><p>Simulate model for a specified time between two time points.</p>
<p>Simulates a time series using the Gillespie algorithm.</p>
<p>Saves a dictionary containing model state history, with keys=times and
values=Model objects with model snapshot at that time point under this
model’s history attribute.</p>
<p>Arguments:
t0 – initial time point to start simulation at (number &gt;= 0)
tf – initial time point to end simulation at (number &gt;= 0)</p>
<p>Keyword arguments:
time_sampling – how many events to skip before saving a snapshot of the</p>
<blockquote>
<div><p>system state (saves all by default), if &lt;0, saves only final state
(int, default 0)</p>
</div></blockquote>
<dl class="simple">
<dt>host_sampling – how many hosts to skip before saving one in a snapshot</dt><dd><p>of the system state (saves all by default) (int &gt;= 0, default 0)</p>
</dd>
<dt>vector_sampling – how many vectors to skip before saving one in a</dt><dd><p>snapshot of the system state (saves all by default)
(int &gt;= 0, default 0)</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="runreplicates">
<h4>runReplicates<a class="headerlink" href="#runreplicates" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">runReplicates</span><span class="p">(</span><span class="n">t0</span><span class="p">,</span><span class="n">tf</span><span class="p">,</span><span class="n">replicates</span><span class="p">,</span><span class="n">host_sampling</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="n">vector_sampling</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</pre></div>
</div>
<p>Simulate replicates of a model, save only end results.</p>
<p>Simulates replicates of a time series using the Gillespie algorithm.</p>
<p>Saves a dictionary containing model end state state, with keys=times and
values=Model objects with model snapshot. The time is the final
timepoint.</p>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>t0 – initial time point to start simulation at (number &gt;= 0)</p></li>
<li><p>tf – initial time point to end simulation at (number &gt;= 0)</p></li>
<li><p>replicates – how many replicates to simulate (int &gt;= 1)</p></li>
</ul>
<p><em>Keyword arguments:</em></p>
<ul class="simple">
<li><p>host_sampling – how many hosts to skip before saving one in a snapshot
of the system state (saves all by default) (int &gt;= 0, default 0)</p></li>
<li><p>vector_sampling – how many vectors to skip before saving one in a
snapshot of the system state (saves all by default)
(int &gt;= 0, default 0)</p></li>
<li><p>**kwargs – additional arguents for joblib multiprocessing</p></li>
</ul>
<p>Returns:
List of Model objects with the final snapshots</p>
</section>
<section id="runparamsweep">
<h4>runParamSweep<a class="headerlink" href="#runparamsweep" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">runParamSweep</span><span class="p">(</span>
<span class="n">t0</span><span class="p">,</span><span class="n">tf</span><span class="p">,</span><span class="n">setup_id</span><span class="p">,</span>
<span class="n">param_sweep_dic</span><span class="o">=</span><span class="p">{},</span><span class="n">pop_ids_param_sweep</span><span class="o">=</span><span class="p">[],</span>
<span class="n">host_population_size_sweep</span><span class="o">=</span><span class="p">{},</span> <span class="n">vector_population_size_sweep</span><span class="o">=</span><span class="p">{},</span>
<span class="n">host_migration_sweep_dic</span><span class="o">=</span><span class="p">{},</span> <span class="n">vector_migration_sweep_dic</span><span class="o">=</span><span class="p">{},</span>
<span class="n">host_host_population_contact_sweep_dic</span><span class="o">=</span><span class="p">{},</span>
<span class="n">host_vector_population_contact_sweep_dic</span><span class="o">=</span><span class="p">{},</span>
<span class="n">vector_host_population_contact_sweep_dic</span><span class="o">=</span><span class="p">{},</span>
<span class="n">replicates</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="n">host_sampling</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="n">vector_sampling</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="n">n_cores</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
<span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</pre></div>
</div>
<p>Simulate a parameter sweep with a model, save only end results.</p>
<p>Simulates variations of a time series using the Gillespie algorithm.</p>
<p>Saves a dictionary containing model end state state, with keys=times and
values=Model objects with model snapshot. The time is the final
timepoint.</p>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>t0 – initial time point to start simulation at (number &gt;= 0)</p></li>
<li><p>tf – initial time point to end simulation at (number &gt;= 0)</p></li>
<li><p>setup_id – ID of setup to be assigned (String)</p></li>
</ul>
<p><em>Keyword Arguments:</em></p>
<ul class="simple">
<li><p>param_sweep_dic – dictionary with keys=parameter names (attributes of
Setup), values=list of values for parameter (list, class of elements
depends on parameter)</p></li>
<li><p>host_population_size_sweep – dictionary with keys=population IDs
(Strings), values=list of values with host population sizes
(must be greater than original size set for each population, list of
numbers)</p></li>
<li><p>vector_population_size_sweep – dictionary with keys=population IDs
(Strings), values=list of values with vector population sizes
(must be greater than original size set for each population, list of
numbers)</p></li>
<li><p>host_migration_sweep_dic – dictionary with keys=population IDs of
origin and destination, separated by a colon ‘;’ (Strings),
values=list of values (list of numbers)</p></li>
<li><p>vector_migration_sweep_dic – dictionary with keys=population IDs of
origin and destination, separated by a colon ‘;’ (Strings),
values=list of values (list of numbers)</p></li>
<li><p>host_host_population_contact_sweep_dic – dictionary with keys=population IDs
of origin and destination, separated by a colon ‘;’ (Strings),
values=list of values (list of numbers)</p></li>
<li><p>host_vector_population_contact_sweep_dic – dictionary with keys=population
IDs of origin and destination, separated by a colon ‘;’ (Strings),
values=list of values (list of numbers)</p></li>
<li><p>vector_host_population_contact_sweep_dic – dictionary with keys=population
IDs of origin and destination, separated by a colon ‘;’ (Strings),
values=list of values (list of numbers)</p></li>
<li><p>replicates – how many replicates to simulate (int &gt;= 1)</p></li>
<li><p>host_sampling – how many hosts to skip before saving one in a snapshot
of the system state (saves all by default) (int &gt;= 0, default 0)</p></li>
<li><p>vector_sampling – how many vectors to skip before saving one in a
snapshot of the system state (saves all by default)
(int &gt;= 0, default 0)</p></li>
<li><p>n_cores – number of cores to parallelize file export across, if 0, all
cores available are used (default 0; int &gt;= 0)</p></li>
<li><p>**kwargs – additional arguents for joblib multiprocessing</p></li>
</ul>
<p><em>Returns:</em></p>
<ul class="simple">
<li><p>DataFrame with parameter combinations, list of Model objects with the
final snapshots</p></li>
</ul>
</section>
<section id="copystate">
<h4>copyState<a class="headerlink" href="#copystate" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">copyState</span><span class="p">(</span><span class="n">host_sampling</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="n">vector_sampling</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
<p>Returns a slimmed-down representation of the current model state.</p>
<p><em>Keyword arguments:</em></p>
<ul class="simple">
<li><p>host_sampling – how many hosts to skip before saving one in a snapshot
of the system state (saves all by default) (int &gt;= 0, default 0)</p></li>
<li><p>vector_sampling – how many vectors to skip before saving one in a
snapshot of the system state (saves all by default)
(int &gt;= 0, default 0)</p></li>
</ul>
<p><em>Returns:</em>
Model object with current population host and vector lists.</p>
</section>
<section id="deepcopy">
<h4>deepCopy<a class="headerlink" href="#deepcopy" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">deepCopy</span><span class="p">()</span>
</pre></div>
</div>
<p>Returns a full copy of the current model with inner references.</p>
<p><em>Returns:</em>
copied Model object</p>
</section>
<section id="savetodataframe">
<h4>saveToDataFrame<a class="headerlink" href="#savetodataframe" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">saveToDataFrame</span><span class="p">(</span><span class="n">save_to_file</span><span class="p">,</span><span class="n">n_cores</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</pre></div>
</div>
<p>Save status of model to dataframe, write to file location given.</p>
<p>Creates a pandas Dataframe in long format with the given model history,
with one host or vector per simulation time in each row, and columns:</p>
<ul class="simple">
<li><p>Time - simulation time of entry</p></li>
<li><p>Population - ID of this host/vector’s population</p></li>
<li><p>Organism - host/vector</p></li>
<li><p>ID - ID of host/vector</p></li>
<li><p>Pathogens - all genomes present in this host/vector separated by ;</p></li>
<li><p>Protection - all genomes present in this host/vector separated by ;</p></li>
<li><p>Alive - whether host/vector is alive at this time, True/False</p></li>
</ul>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>save_to_file – file path and name to save model data under (String)</p></li>
</ul>
<p><em>Keyword Arguments:</em></p>
<ul class="simple">
<li><p>n_cores – number of cores to parallelize file export across, if 0, all
cores available are used (default 0; int)</p></li>
<li><p>**kwargs – additional arguents for joblib multiprocessing</p></li>
</ul>
<p><em>Returns:</em></p>
<ul class="simple">
<li><p>pandas dataframe with model history as described above</p></li>
</ul>
</section>
<section id="getcompositiondata">
<h4>getCompositionData<a class="headerlink" href="#getcompositiondata" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">getCompositionData</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">populations</span><span class="o">=</span><span class="p">[],</span> <span class="n">type_of_composition</span><span class="o">=</span><span class="s1">&#39;Pathogens&#39;</span><span class="p">,</span>
    <span class="n">hosts</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">vectors</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">num_top_sequences</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">track_specific_sequences</span><span class="o">=</span><span class="p">[],</span> <span class="n">genomic_positions</span><span class="o">=</span><span class="p">[],</span>
    <span class="n">count_individuals_based_on_model</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">save_data_to_file</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">n_cores</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
    <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</pre></div>
</div>
<p>Create dataframe with counts for pathogen genomes or resistance.</p>
<p>Creates a pandas Dataframe with dynamics of the pathogen strains or
protection sequences across selected populations in the model,
with one time point in each row and columns for pathogen genomes or
protection sequences.</p>
<p>Of note: sum of totals for all sequences in one time point does not
necessarily equal the number of infected hosts and/or vectors, given
multiple infections in the same host/vector are counted separately.</p>
<p><em>Keyword Arguments:</em></p>
<ul class="simple">
<li><p>data – dataframe with model history as produced by saveToDf function;
if None, computes this dataframe and saves it under
‘raw_data_’+save_data_to_file (DataFrame, default None)</p></li>
<li><p>populations – IDs of populations to include in analysis; if empty, uses
all populations in model (default empty list; list of Strings)</p></li>
<li><p>type_of_composition – field of data to count totals of, can be either
‘Pathogens’ or ‘Protection’ (default ‘Pathogens’; String)</p></li>
<li><p>hosts – whether to count hosts (default True, Boolean)</p></li>
<li><p>vectors – whether to count vectors (default False, Boolean)</p></li>
<li><p>num_top_sequences – how many sequences to count separately and include
as columns, remainder will be counted under column “Other”; if &lt;0,
includes all genomes in model (default -1; int)</p></li>
<li><p>track_specific_sequences – contains specific sequences to have
as a separate column if not part of the top num_top_sequences
sequences (default empty list; list of Strings)</p></li>
<li><p>genomic_positions – list in which each element is a list with loci
positions to extract (e.g. genomic_positions=[ [0,3], [5,6] ]
extracts positions 0, 1, 2, and 5 from each genome); if empty, takes
full genomes(default empty list; list of lists of int)</p></li>
<li><p>count_individuals_based_on_model – Model object with populations and
fitness functions used to evaluate the most fit pathogen genome in
each host/vector in order to count only a single pathogen per
host/vector, asopposed to all pathogens within each host/vector; if
None, counts all pathogens (default None; None or Model)</p></li>
<li><p>save_data_to_file – file path and name to save model data under, no
saving occurs if empty string (default ‘’; String)</p></li>
<li><p>n_cores – number of cores to parallelize processing across, if 0, all
cores available are used (default 0; int)</p></li>
<li><p>**kwargs – additional arguents for joblib multiprocessing</p></li>
</ul>
<p><em>Returns:</em></p>
<ul class="simple">
<li><p>pandas dataframe with model sequence composition dynamics as described
above</p></li>
</ul>
</section>
<section id="getpathogens">
<h4>getPathogens<a class="headerlink" href="#getpathogens" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">getPathogens</span><span class="p">(</span><span class="n">dat</span><span class="p">,</span> <span class="n">save_to_file</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Create Dataframe with counts for all pathogen genomes in data.</p>
<p>Returns sorted pandas Dataframe with counts for occurrences of all pathogen
genomes in data passed.</p>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>data – dataframe with model history as produced by saveToDf function</p></li>
</ul>
<p><em>Keyword Arguments:</em></p>
<ul class="simple">
<li><p>save_to_file – file path and name to save model data under, no saving
occurs if empty string (default ‘’; String)</p></li>
</ul>
<p><em>Returns:</em></p>
<ul class="simple">
<li><p>pandas dataframe with Series as described above</p></li>
</ul>
</section>
<section id="getprotections">
<h4>getProtections<a class="headerlink" href="#getprotections" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">getProtections</span><span class="p">(</span><span class="n">dat</span><span class="p">,</span> <span class="n">save_to_file</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Create Dataframe with counts for all protection sequences in data.</p>
<p>Returns sorted pandas Dataframe with counts for occurrences of all
protection sequences in data passed.</p>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>data – dataframe with model history as produced by saveToDf function</p></li>
</ul>
<p><em>Keyword Arguments:</em></p>
<ul class="simple">
<li><p>save_to_file – file path and name to save model data under, no saving</p></li>
<li><p>occurs if empty string (default ‘’; String)</p></li>
</ul>
<p><em>Returns:</em></p>
<ul class="simple">
<li><p>pandas dataframe with Series as described above</p></li>
</ul>
</section>
<section id="populationsplot">
<h4>populationsPlot<a class="headerlink" href="#populationsplot" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">populationsPlot</span><span class="p">(</span>
<span class="n">file_name</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">compartment</span><span class="o">=</span><span class="s1">&#39;Infected&#39;</span><span class="p">,</span>
<span class="n">hosts</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">vectors</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">num_top_populations</span><span class="o">=</span><span class="mi">7</span><span class="p">,</span>
<span class="n">track_specific_populations</span><span class="o">=</span><span class="p">[],</span> <span class="n">save_data_to_file</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
<span class="n">x_label</span><span class="o">=</span><span class="s1">&#39;Time&#39;</span><span class="p">,</span> <span class="n">y_label</span><span class="o">=</span><span class="s1">&#39;Hosts&#39;</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span>
<span class="n">palette</span><span class="o">=</span><span class="n">CB_PALETTE</span><span class="p">,</span> <span class="n">stacked</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<p>Create plot with aggregated totals per population across time.</p>
<p>Creates a line or stacked line plot with dynamics of a compartment
across populations in the model, with one line for each population.</p>
<p>A host or vector is considered part of the recovered compartment
if it has protection sequences of any kind and is not infected.</p>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>file_name – file path, name, and extension to save plot under (String)</p></li>
<li><p>data – dataframe with model history as produced by saveToDf function
(DataFrame)</p></li>
</ul>
<p><em>Keyword Arguments:</em></p>
<ul class="simple">
<li><p>compartment – subset of hosts/vectors to count totals of, can be either
‘Naive’,’Infected’,’Recovered’, or ‘Dead’ (default ‘Infected’;
String)</p></li>
<li><p>hosts – whether to count hosts (default True, Boolean)</p></li>
<li><p>vectors – whether to count vectors (default False, Boolean)</p></li>
<li><p>num_top_populations – how many populations to count separately and
include as columns, remainder will be counted under column “Other”;
if &lt;0, includes all populations in model (default 7; int)</p></li>
<li><p>track_specific_populations – contains IDs of specific populations to
have as a separate column if not part of the top num_top_populations
populations (list of Strings)</p></li>
<li><p>save_data_to_file – file path and name to save model plot data under,
no saving occurs if empty string (default ‘’; String)</p></li>
<li><p>x_label – X axis title (default ‘Time’, String)</p></li>
<li><p>y_label – Y axis title (default ‘Hosts’, String)</p></li>
<li><p>legend_title – legend title (default ‘Population’, String)</p></li>
<li><p>legend_values – labels for each trace, if empty list, uses population
IDs (default empty list, list of Strings)</p></li>
<li><p>figsize – dimensions of figure (default (8,4), array-like of two ints)</p></li>
<li><p>dpi – figure resolution (default 200, int)</p></li>
<li><p>palette – color palette to use for traces (default CB_PALETTE, list of
color Strings)</p></li>
<li><p>stacked – whether to draw a regular line plot or a stacked one (default
False, Boolean)</p></li>
</ul>
<p><em>Returns:</em></p>
<ul class="simple">
<li><p>axis object for plot with model population dynamics as described above</p></li>
</ul>
</section>
<section id="compartmentplot">
<h4>compartmentPlot<a class="headerlink" href="#compartmentplot" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">compartmentPlot</span><span class="p">(</span>
<span class="n">file_name</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">populations</span><span class="o">=</span><span class="p">[],</span> <span class="n">hosts</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">vectors</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="n">save_data_to_file</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">x_label</span><span class="o">=</span><span class="s1">&#39;Time&#39;</span><span class="p">,</span> <span class="n">y_label</span><span class="o">=</span><span class="s1">&#39;Hosts&#39;</span><span class="p">,</span>
<span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="n">CB_PALETTE</span><span class="p">,</span> <span class="n">stacked</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<p>Create plot with number of naive, infected, recovered, dead hosts/vectors vs.
time.</p>
<p>Creates a line or stacked line plot with dynamics of all compartments
(naive, infected, recovered, dead) across selected populations in the
model, with one line for each compartment.</p>
<p>A host or vector is considered recovered if it has protection sequences of any
kind and is not infected.</p>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>file_name – file path, name, and extension to save plot under (String)</p></li>
<li><p>data – dataframe with model history as produced by saveToDf function
(DataFrame)</p></li>
</ul>
<p><em>Keyword Arguments:</em></p>
<ul class="simple">
<li><p>populations – IDs of populations to include in analysis; if empty, uses</p></li>
<li><p>all populations in model (default empty list; list of Strings)</p></li>
<li><p>hosts – whether to count hosts (default True, Boolean)</p></li>
<li><p>vectors – whether to count vectors (default False, Boolean)</p></li>
<li><p>save_data_to_file – file path and name to save model data under, no saving</p></li>
<li><p>occurs if empty string (default ‘’; String)</p></li>
<li><p>x_label – X axis title (default ‘Time’, String)</p></li>
<li><p>y_label – Y axis title (default ‘Hosts’, String)</p></li>
<li><p>legend_title – legend title (default ‘Population’, String)</p></li>
<li><p>legend_values – labels for each trace, if empty list, uses population
IDs (default empty list, list of Strings)</p></li>
<li><p>figsize – dimensions of figure (default (8,4), array-like of two ints)</p></li>
<li><p>dpi – figure resolution (default 200, int)</p></li>
<li><p>palette – color palette to use for traces (default CB_PALETTE, list of
color Strings)</p></li>
<li><p>stacked – whether to draw a regular line plot or a stacked one (default
False, Boolean)</p></li>
</ul>
<p><em>Returns:</em></p>
<ul class="simple">
<li><p>axis object for plot with model compartment dynamics as described above</p></li>
</ul>
</section>
<section id="compositionplot">
<h4>compositionPlot<a class="headerlink" href="#compositionplot" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">compositionPlot</span><span class="p">(</span>
<span class="n">file_name</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">composition_dataframe</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">populations</span><span class="o">=</span><span class="p">[],</span>
<span class="n">type_of_composition</span><span class="o">=</span><span class="s1">&#39;Pathogens&#39;</span><span class="p">,</span> <span class="n">hosts</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">vectors</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="n">num_top_sequences</span><span class="o">=</span><span class="mi">7</span><span class="p">,</span> <span class="n">track_specific_sequences</span><span class="o">=</span><span class="p">[],</span>
<span class="n">genomic_positions</span><span class="o">=</span><span class="p">[],</span> <span class="n">count_individuals_based_on_model</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<span class="n">remove_legend</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">population_fraction</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="n">save_data_to_file</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">x_label</span><span class="o">=</span><span class="s1">&#39;Time&#39;</span><span class="p">,</span> <span class="n">y_label</span><span class="o">=</span><span class="s1">&#39;Infections&#39;</span><span class="p">,</span>
<span class="n">legend_title</span><span class="o">=</span><span class="s1">&#39;Genotype&#39;</span><span class="p">,</span> <span class="n">legend_values</span><span class="o">=</span><span class="p">[],</span>
<span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="n">CB_PALETTE</span><span class="p">,</span> <span class="n">stacked</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</pre></div>
</div>
<p>Create plot with counts for pathogen genomes or resistance vs. time.</p>
<p>Creates a line or stacked line plot with dynamics of the pathogen
strains or protection sequences across selected populations in the
model, with one line for each pathogen genome or protection sequence
being shown.s</p>
<p>Of note: sum of totals for all sequences in one time point does not
necessarily equal the number of infected hosts and/or vectors, given
multiple infections in the same host/vector are counted separately.</p>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>file_name – file path, name, and extension to save plot under (String)</p></li>
<li><p>data – dataframe with model history as produced by saveToDf function</p></li>
</ul>
<p><em>Keyword Arguments:</em></p>
<ul class="simple">
<li><p>composition_dataframe – output of compositionDf() if already computed
(Pandas DataFrame, None by default)</p></li>
<li><p>populations – IDs of populations to include in analysis; if empty, uses
all populations in model (default empty list; list of Strings)</p></li>
<li><p>type_of_composition – field of data to count totals of, can be either
‘Pathogens’ or ‘Protection’ (default ‘Pathogens’; String)</p></li>
<li><p>hosts – whether to count hosts (default True, Boolean)</p></li>
<li><p>vectors – whether to count vectors (default False, Boolean)</p></li>
<li><p>num_top_sequences – how many sequences to count separately and include
as columns, remainder will be counted under column “Other”; if &lt;0,
includes all genomes in model (default 7; int)</p></li>
<li><p>track_specific_sequences – contains specific sequences to have
as a separate column if not part of the top num_top_sequences
sequences (list of Strings)</p></li>
<li><p>genomic_positions – list in which each element is a list with loci
positions to extract (e.g. genomic_positions=[ [0,3], [5,6] ] extracts
positions 0, 1, 2, and 5 from each genome); if empty, takes full genomes
(default empty list; list of lists of int)</p></li>
<li><p>count_individuals_based_on_model – Model object with populations and
fitness functions used to evaluate the most fit pathogen genome in each
host/vector in order to count only a single pathogen per host/vector, as
opposed to all pathogens within each host/vector; if None, counts all
pathogens (default None; None or Model)</p></li>
<li><p>save_data_to_file – file path and name to save model data under, no saving
occurs if empty string (default ‘’; String)</p></li>
<li><p>x_label – X axis title (default ‘Time’, String)</p></li>
<li><p>y_label – Y axis title (default ‘Hosts’, String)</p></li>
<li><p>legend_title – legend title (default ‘Population’, String)</p></li>
<li><p>legend_values – labels for each trace, if empty list, uses population
IDs (default empty list, list of Strings)</p></li>
<li><p>figsize – dimensions of figure (default (8,4), array-like of two ints)</p></li>
<li><p>dpi – figure resolution (default 200, int)</p></li>
<li><p>palette – color palette to use for traces (default CB_PALETTE, list of
color Strings)</p></li>
<li><p>stacked – whether to draw a regular line plot instead of a stacked one
(default False, Boolean).</p></li>
<li><p>remove_legend – whether to print the sequences on the figure legend instead
of printing them on a separate csv file (default True; Boolean)</p></li>
<li><p>population_fraction – whether to graph fractions of pathogen population
instead of pathogen counts (default False, Boolean)</p></li>
<li><p>**kwargs – additional arguents for joblib multiprocessing</p></li>
</ul>
<p><em>Returns:</em></p>
<ul class="simple">
<li><p>axis object for plot with model sequence composition dynamics as
described</p></li>
</ul>
</section>
<section id="clustermap">
<h4>clustermap<a class="headerlink" href="#clustermap" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">clustermap</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">num_top_sequences</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span>
<span class="n">track_specific_sequences</span><span class="o">=</span><span class="p">[],</span> <span class="n">seq_names</span><span class="o">=</span><span class="p">[],</span> <span class="n">n_cores</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;weighted&#39;</span><span class="p">,</span>
<span class="n">metric</span><span class="o">=</span><span class="s1">&#39;euclidean&#39;</span><span class="p">,</span><span class="n">save_data_to_file</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">legend_title</span><span class="o">=</span><span class="s1">&#39;Distance&#39;</span><span class="p">,</span>
<span class="n">legend_values</span><span class="o">=</span><span class="p">[],</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">10</span><span class="p">),</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">color_map</span><span class="o">=</span><span class="n">DEF_CMAP</span><span class="p">)</span>
</pre></div>
</div>
<p>Create a heatmap and dendrogram for pathogen genomes in data passed.</p>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>file_name – file path, name, and extension to save plot under (String)</p></li>
<li><p>data – dataframe with model history as produced by saveToDf function</p></li>
</ul>
<p><em>Keyword arguments:</em></p>
<ul class="simple">
<li><p>num_top_sequences – how many sequences to include in matrix; if &lt;0,
includes all genomes in data passed (default -1; int)</p></li>
<li><p>track_specific_sequences – contains specific sequences to include in
matrixif not part of the top num_top_sequences sequences (default
empty list; list of Strings)</p></li>
<li><p>seq_names – list with names to be used for sequence labels in matrix
must be of same length as number of sequences to be displayed; if
empty, uses sequences themselves (default empty list; list of
Strings)</p></li>
<li><p>n_cores – number of cores to parallelize distance compute across, if 0,
all cores available are used (default 0; int)</p></li>
<li><p>method – clustering algorithm to use with seaborn clustermap (default
‘weighted’; String)</p></li>
<li><p>metric – distance metric to use with seaborn clustermap (default
‘euclidean’; String)</p></li>
<li><p>save_data_to_file – file path and name to save model data under, no
saving occurs if empty string (default ‘’; String)</p></li>
<li><p>legend_title – legend title (default ‘Distance’, String)</p></li>
<li><p>figsize – dimensions of figure (default (8,4), array-like of two ints)</p></li>
<li><p>dpi – figure resolution (default 200, int)</p></li>
<li><p>color_map – color map to use for traces (default DEF_CMAP, cmap object)</p></li>
</ul>
<p><em>Returns:</em></p>
<ul class="simple">
<li><p>figure object for plot with heatmap and dendrogram as described</p></li>
</ul>
</section>
<section id="pathogendistancehistory">
<h4>pathogenDistanceHistory<a class="headerlink" href="#pathogendistancehistory" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">pathogenDistanceHistory</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">samples</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_top_sequences</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span>
<span class="n">track_specific_sequences</span><span class="o">=</span><span class="p">[],</span> <span class="n">seq_names</span><span class="o">=</span><span class="p">[],</span> <span class="n">n_cores</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">save_to_file</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Create a long-format dataframe with pairwise distances for pathogen
genomes in data passed for different time points.</p>
<p><em>Arguments:</em>
data – dataframe with model history as produced by saveToDf function</p>
<p><em>Keyword Arguments:</em></p>
<ul class="simple">
<li><p>samples – how many timepoints to uniformly sample from the total
timecourse; if &lt;0, takes all timepoints (default -1; int)</p></li>
<li><p>num_top_sequences – how many sequences to include in matrix; if &lt;0,
includes all genomes in data passed (default -1; int)</p></li>
<li><p>track_specific_sequences – contains specific sequences to include in
matrixif not part of the top num_top_sequences sequences (default
empty list; list of Strings)</p></li>
<li><p>seq_names – list with names to be used for sequence labels in matrix
must be of same length as number of sequences to be displayed; if
empty, uses sequences themselves (default empty list; list of
Strings)</p></li>
<li><p>n_cores – number of cores to parallelize distance compute across, if 0,
all cores available are used (default 0; int)</p></li>
<li><p>method – clustering algorithm to use with seaborn clustermap (default
‘weighted’; String)</p></li>
<li><p>metric – distance metric to use with seaborn clustermap (default
‘euclidean’; String)</p></li>
<li><p>save_data_to_file – file path and name to save model data under, no
saving occurs if empty string (default ‘’; String)</p></li>
</ul>
<p><em>Returns:</em></p>
<ul class="simple">
<li><p>long-format Pandas dataframe with pairwise distances for pathogen
genomes in data passed for different time points.</p></li>
</ul>
</section>
<section id="getgenometimes">
<h4>getGenomeTimes<a class="headerlink" href="#getgenometimes" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">getGenomeTimes</span><span class="p">(</span>
      <span class="n">data</span><span class="p">,</span> <span class="n">samples</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_top_sequences</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span> <span class="n">track_specific_sequences</span><span class="o">=</span><span class="p">[],</span>
      <span class="n">seq_names</span><span class="o">=</span><span class="p">[],</span> <span class="n">n_cores</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">save_to_file</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Create DataFrame with times genomes first appeared during simulation.</p>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>data – dataframe with model history as produced by saveToDf function</p></li>
</ul>
<p><em>Keyword arguments:</em></p>
<ul class="simple">
<li><p>samples – how many timepoints to uniformly sample from the total
timecourse; if &lt;0, takes all timepoints (default 1; int)</p></li>
<li><p>save_to_file – file path and name to save model data under, no saving
occurs if empty string (default ‘’; String)</p></li>
<li><p>n_cores – number of cores to parallelize across, if 0, all cores
available are used (default 0; int)</p></li>
</ul>
<p><em>Returns:</em></p>
<ul class="simple">
<li><p>pandas dataframe with genomes and times as described above</p></li>
</ul>
</section>
<section id="newpopulation">
<h4>newPopulation<a class="headerlink" href="#newpopulation" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">newPopulation</span><span class="p">(</span><span class="nb">id</span><span class="p">,</span> <span class="n">setup_name</span><span class="p">,</span> <span class="n">num_hosts</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">num_vectors</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
</pre></div>
</div>
<p>Create a new Population object with setup parameters.</p>
<p>If population ID is already in use, appends _2 to it</p>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>id – unique identifier for this population in the model (String)</p></li>
<li><p>setup_name – setup object with parameters for this population (Setup)</p></li>
</ul>
<p><em>Keyword Arguments:</em></p>
<ul class="simple">
<li><p>num_hosts – number of hosts to initialize population with (default 100;
int)</p></li>
<li><p>num_vectors – number of hosts to initialize population with (default
100; int)</p></li>
</ul>
</section>
<section id="linkpopulationshostmigration">
<h4>linkPopulationsHostMigration<a class="headerlink" href="#linkpopulationshostmigration" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">linkPopulationsHostMigration</span><span class="p">(</span><span class="n">pop1_id</span><span class="p">,</span> <span class="n">pop2_id</span><span class="p">,</span> <span class="n">rate</span><span class="p">)</span>
</pre></div>
</div>
<p>Set host migration rate from one population towards another.</p>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>pop1_id – origin population for which migration rate will be specified
(String)</p></li>
<li><p>pop1_id – destination population for which migration rate will be
specified (String)</p></li>
<li><p>rate – migration rate from one population to the neighbor; events/time
(number &gt;= 0)</p></li>
</ul>
</section>
<section id="linkpopulationsvectormigration">
<h4>linkPopulationsVectorMigration<a class="headerlink" href="#linkpopulationsvectormigration" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">linkPopulationsVectorMigration</span><span class="p">(</span><span class="n">pop1_id</span><span class="p">,</span> <span class="n">pop2_id</span><span class="p">,</span> <span class="n">rate</span><span class="p">)</span>
</pre></div>
</div>
<p>Set vector migration rate from one population towards another.</p>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>pop1_id – origin population for which migration rate will be specified
(String)</p></li>
<li><p>pop1_id – destination population for which migration rate will be
specified (String)</p></li>
<li><p>rate – migration rate from one population to the neighbor; events/time
(number &gt;= 0)</p></li>
</ul>
</section>
<section id="linkpopulationshosthostcontact">
<h4>linkPopulationsHostHostContact<a class="headerlink" href="#linkpopulationshosthostcontact" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">linkPopulationsHostHostContact</span><span class="p">(</span><span class="n">pop1_id</span><span class="p">,</span> <span class="n">pop2_id</span><span class="p">,</span> <span class="n">rate</span><span class="p">)</span>
</pre></div>
</div>
<p>Set host-host inter-population contact rate from one population towards another.</p>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>pop1_id – origin population for which migration rate will be specified
(String)</p></li>
<li><p>pop1_id – destination population for which migration rate will be
specified (String)</p></li>
<li><p>rate – migration rate from one population to the neighbor; events/time
(number &gt;= 0)</p></li>
</ul>
</section>
<section id="linkpopulationshostvectorcontact">
<h4>linkPopulationsHostVectorContact<a class="headerlink" href="#linkpopulationshostvectorcontact" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">linkPopulationsHostVectorContact</span><span class="p">(</span><span class="n">pop1_id</span><span class="p">,</span> <span class="n">pop2_id</span><span class="p">,</span> <span class="n">rate</span><span class="p">)</span>
</pre></div>
</div>
<p>Set host-vector inter-population contact rate from one population to another.</p>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>pop1_id – origin population for which migration rate will be specified
(String)</p></li>
<li><p>pop1_id – destination population for which migration rate will be
specified (String)</p></li>
<li><p>rate – migration rate from one population to the neighbor; events/time
(number &gt;= 0)</p></li>
</ul>
</section>
<section id="linkpopulationsvectorhostcontact">
<h4>linkPopulationsVectorHostContact<a class="headerlink" href="#linkpopulationsvectorhostcontact" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">linkPopulationsVectorHostContact</span><span class="p">(</span><span class="n">pop1_id</span><span class="p">,</span> <span class="n">pop2_id</span><span class="p">,</span> <span class="n">rate</span><span class="p">)</span>
</pre></div>
</div>
<p>Set vector-host inter-population contact rate from one population to another.</p>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>pop1_id – origin population for which migration rate will be specified
(String)</p></li>
<li><p>pop1_id – destination population for which migration rate will be
specified (String)</p></li>
<li><p>rate – migration rate from one population to the neighbor; events/time
(number &gt;= 0)</p></li>
</ul>
</section>
<section id="createinterconnectedpopulations">
<h4>createInterconnectedPopulations<a class="headerlink" href="#createinterconnectedpopulations" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">createInterconnectedPopulations</span><span class="p">(</span>
      <span class="n">num_populations</span><span class="p">,</span> <span class="n">id_prefix</span><span class="p">,</span> <span class="n">setup_name</span><span class="p">,</span>
      <span class="n">host_migration_rate</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">vector_migration_rate</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
      <span class="n">host_host_contact_rate</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
      <span class="n">host_vector_contact_rate</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">vector_host_contact_rate</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
      <span class="n">num_hosts</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">num_vectors</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
</pre></div>
</div>
<p>Create new populations, link all of them to each other.</p>
<p>All populations in this cluster are linked with the same migration rate,
starting number of hosts and vectors, and setup parameters. Their IDs
are numbered onto prefix given as ‘id_prefix_0’, ‘id_prefix_1’,
‘id_prefix_2’, etc.</p>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>num_populations – number of populations to be created (int)</p></li>
<li><p>id_prefix – prefix for IDs to be used for this population in the model,
(String)</p></li>
<li><p>setup_name – setup object with parameters for all populations (Setup)</p></li>
</ul>
<p><em>Keyword arguments:</em></p>
<ul class="simple">
<li><p>host_migration_rate – host migration rate between populations;
events/time (default 0; number &gt;= 0)</p></li>
<li><p>vector_migration_rate – vector migration rate between populations;
events/time (default 0; number &gt;= 0)</p></li>
<li><p>host_host_contact_rate – host-host inter-population contact rate between
populations; events/time (default 0; number &gt;= 0)</p></li>
<li><p>host_vector_contact_rate – host-vector inter-population contact rate between
populations; events/time (default 0; number &gt;= 0)</p></li>
<li><p>vector_host_contact_rate – vector-host inter-population contact rate between
populations; events/time (default 0; number &gt;= 0)</p></li>
<li><p>num_hosts – number of hosts to initialize population with (default 100;
int)</p></li>
<li><p>num_vectors – number of hosts to initialize population with (default
100; int)</p></li>
</ul>
</section>
<section id="newhostgroup">
<h4>newHostGroup<a class="headerlink" href="#newhostgroup" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">newHostGroup</span><span class="p">(</span><span class="n">pop_id</span><span class="p">,</span> <span class="n">group_id</span><span class="p">,</span> <span class="n">num_hosts</span><span class="p">,</span> <span class="n">healthy</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<p>Return a list of random (healthy or any) hosts in population.</p>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>pop_id – ID of population to be modified (String)</p></li>
<li><p>group_id – ID to call this group by (String)</p></li>
<li><p>num_vectors – number of vectors to be sampled randomly (int)</p></li>
</ul>
<p><em>Keyword Arguments:</em></p>
<ul class="simple">
<li><p>healthy – whether to sample healthy hosts only (default True; Boolean)</p></li>
</ul>
<p><em>Returns:</em></p>
<ul class="simple">
<li><p>list containing sampled hosts</p></li>
</ul>
</section>
<section id="newvectorgroup">
<h4>newVectorGroup<a class="headerlink" href="#newvectorgroup" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">newVectorGroup</span><span class="p">(</span><span class="n">pop_id</span><span class="p">,</span> <span class="n">group_id</span><span class="p">,</span> <span class="n">num_vectors</span><span class="p">,</span> <span class="n">healthy</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<p>Return a list of random (healthy or any) vectors in population.</p>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>pop_id – ID of population to be modified (String)</p></li>
<li><p>group_id – ID to call this group by (String)</p></li>
<li><p>num_vectors – number of vectors to be sampled randomly (int)</p></li>
</ul>
<p><em>Keyword Arguments:</em></p>
<ul class="simple">
<li><p>healthy – whether to sample healthy vectors only (default True;
Boolean)</p></li>
</ul>
<p><em>Returns:</em></p>
<ul class="simple">
<li><p>list containing sampled vectors</p></li>
</ul>
</section>
<section id="addhosts">
<h4>addHosts<a class="headerlink" href="#addhosts" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">addHosts</span><span class="p">(</span><span class="n">pop_id</span><span class="p">,</span> <span class="n">num_hosts</span><span class="p">)</span>
</pre></div>
</div>
<p>Add a number of healthy hosts to population, return list with them.</p>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>pop_id – ID of population to be modified (String)</p></li>
<li><p>num_hosts – number of hosts to be added (int)</p></li>
</ul>
<p><em>Returns:</em></p>
<ul class="simple">
<li><p>list containing new hosts</p></li>
</ul>
</section>
<section id="addvectors">
<h4>addVectors<a class="headerlink" href="#addvectors" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">addVectors</span><span class="p">(</span><span class="n">pop_id</span><span class="p">,</span> <span class="n">num_vectors</span><span class="p">)</span>
</pre></div>
</div>
<p>Add a number of healthy vectors to population, return list with them.</p>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>pop_id – ID of population to be modified (String)</p></li>
<li><p>num_vectors – number of vectors to be added (int)</p></li>
</ul>
<p><em>Returns:</em></p>
<ul class="simple">
<li><p>list containing new vectors</p></li>
</ul>
</section>
<section id="removehosts">
<h4>removeHosts<a class="headerlink" href="#removehosts" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">removeHosts</span><span class="p">(</span><span class="n">pop_id</span><span class="p">,</span> <span class="n">num_hosts_or_list</span><span class="p">)</span>
</pre></div>
</div>
<p>Remove a number of specified or random hosts from population.</p>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>pop_id – ID of population to be modified (String)</p></li>
<li><p>num_hosts_or_list – number of hosts to be sampled randomly for removal
or list of hosts to be removed, must be hosts in this population
(int or list of Hosts)</p></li>
</ul>
</section>
<section id="removevectors">
<h4>removeVectors<a class="headerlink" href="#removevectors" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">removeVectors</span><span class="p">(</span><span class="n">pop_id</span><span class="p">,</span> <span class="n">num_vectors_or_list</span><span class="p">)</span>
</pre></div>
</div>
<p>Remove a number of specified or random vectors from population.</p>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>pop_id – ID of population to be modified (String)</p></li>
<li><p>num_vectors_or_list – number of vectors to be sampled randomly for
removal or list of vectors to be removed, must be vectors in this
population (int or list of Vectors)</p></li>
</ul>
</section>
<section id="addpathogenstohosts">
<h4>addPathogensToHosts<a class="headerlink" href="#addpathogenstohosts" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">addPathogensToHosts</span><span class="p">(</span><span class="n">pop_id</span><span class="p">,</span> <span class="n">genomes_numbers</span><span class="p">,</span> <span class="n">group_id</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Add specified pathogens to random hosts, optionally from a list.</p>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>pop_id – ID of population to be modified (String)</p></li>
<li><p>genomes_numbers – dictionary containing pathogen genomes to add as keys
and number of hosts each one will be added to as values (dict with
keys=Strings, values=int)</p></li>
</ul>
<p><em>Keyword Arguments:</em></p>
<ul class="simple">
<li><p>group_id – ID of group to sample hosts to sample from, if empty, samples from
whole population (default empty String; String)</p></li>
</ul>
</section>
<section id="addpathogenstovectors">
<h4>addPathogensToVectors<a class="headerlink" href="#addpathogenstovectors" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">addPathogensToVectors</span><span class="p">(</span><span class="n">pop_id</span><span class="p">,</span> <span class="n">genomes_numbers</span><span class="p">,</span> <span class="n">group_id</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Add specified pathogens to random vectors, optionally from a list.</p>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>pop_id – ID of population to be modified (String)</p></li>
<li><p>genomes_numbers – dictionary containing pathogen genomes to add as keys
and number of vectors each one will be added to as values (dict with
keys=Strings, values=int)</p></li>
</ul>
<p><em>Keyword Arguments:</em></p>
<ul class="simple">
<li><p>group_id – ID of group to sample vectors to sample from, if empty, samples
from whole population (default empty String; String)</p></li>
</ul>
</section>
<section id="treathosts">
<h4>treatHosts<a class="headerlink" href="#treathosts" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">treatHosts</span><span class="p">(</span><span class="n">pop_id</span><span class="p">,</span> <span class="n">frac_hosts</span><span class="p">,</span> <span class="n">resistance_seqs</span><span class="p">,</span> <span class="n">group_id</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Treat random fraction of infected hosts against some infection.</p>
<p>Removes all infections with genotypes susceptible to given treatment.
Pathogens are removed if they are missing at least one of the sequences
in resistance_seqs from their genome. Removes this organism from
population infected list and adds to healthy list if appropriate.</p>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>pop_id – ID of population to be modified (String)</p></li>
<li><p>frac_hosts – fraction of hosts considered to be randomly selected
(number between 0 and 1)</p></li>
<li><p>resistance_seqs – contains sequences required for treatment resistance
(list of Strings)</p></li>
</ul>
<p><em>Keyword Arguments:</em></p>
<ul class="simple">
<li><p>group_id – ID of group to sample hosts to sample from, if empty, samples from
whole population (default empty String; String)</p></li>
</ul>
</section>
<section id="treatvectors">
<h4>treatVectors<a class="headerlink" href="#treatvectors" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">treatVectors</span><span class="p">(</span><span class="n">pop_id</span><span class="p">,</span> <span class="n">frac_vectors</span><span class="p">,</span> <span class="n">resistance_seqs</span><span class="p">,</span> <span class="n">group_id</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Treat random fraction of infected vectors agains some infection.</p>
<p>Removes all infections with genotypes susceptible to given treatment.
Pathogens are removed if they are missing at least one of the sequences
in resistance_seqs from their genome. Removes this organism from
population infected list and adds to healthy list if appropriate.</p>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>pop_id – ID of population to be modified (String)</p></li>
<li><p>frac_vectors – fraction of vectors considered to be randomly selected
(number between 0 and 1)</p></li>
<li><p>resistance_seqs – contains sequences required for treatment resistance
(list of Strings)</p></li>
</ul>
<p><em>Keyword Arguments:</em></p>
<ul class="simple">
<li><p>group_id – ID of group to sample vectors to sample from, if empty, samples
from whole population (default empty String; String)</p></li>
</ul>
</section>
<section id="protecthosts">
<h4>protectHosts<a class="headerlink" href="#protecthosts" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">protectHosts</span><span class="p">(</span><span class="n">pop_id</span><span class="p">,</span> <span class="n">frac_hosts</span><span class="p">,</span> <span class="n">protection_sequence</span><span class="p">,</span> <span class="n">group_id</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Protect a random fraction of infected hosts against some infection.</p>
<p>Adds protection sequence specified to a random fraction of the hosts
specified. Does not cure them if they are already infected.</p>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>pop_id – ID of population to be modified (String)</p></li>
<li><p>frac_hosts – fraction of hosts considered to be randomly selected
(number between 0 and 1)</p></li>
<li><p>protection_sequence – sequence against which to protect (String)</p></li>
</ul>
<p><em>Keyword Arguments:</em></p>
<ul class="simple">
<li><p>group_id – ID of group to sample hosts to sample from, if empty, samples from
whole population (default empty String; String)</p></li>
</ul>
</section>
<section id="protectvectors">
<h4>protectVectors<a class="headerlink" href="#protectvectors" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">protectVectors</span><span class="p">(</span><span class="n">pop_id</span><span class="p">,</span> <span class="n">frac_vectors</span><span class="p">,</span> <span class="n">protection_sequence</span><span class="p">,</span> <span class="n">group_id</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Protect a random fraction of infected vectors against some infection.</p>
<p>Adds protection sequence specified to a random fraction of the vectors
specified. Does not cure them if they are already infected.</p>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>pop_id – ID of population to be modified (String)</p></li>
<li><p>frac_vectors – fraction of vectors considered to be randomly selected
(number between 0 and 1)</p></li>
<li><p>protection_sequence – sequence against which to protect (String)</p></li>
</ul>
<p><em>Keyword Arguments:</em></p>
<ul class="simple">
<li><p>group_id – ID of group to sample vectors to sample from, if empty, samples
from whole population (default empty String; String)</p></li>
</ul>
</section>
<section id="wipeprotectionhosts">
<h4>wipeProtectionHosts<a class="headerlink" href="#wipeprotectionhosts" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">wipeProtectionHosts</span><span class="p">(</span><span class="n">pop_id</span><span class="p">,</span> <span class="n">group_id</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Removes all protection sequences from hosts.</p>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>pop_id – ID of population to be modified (String)</p></li>
</ul>
<p><em>Keyword Arguments:</em></p>
<ul class="simple">
<li><p>group_id – ID of group to sample hosts to sample from, if empty, takes
whole population (default empty String; String)</p></li>
</ul>
</section>
<section id="wipeprotectionvectors">
<h4>wipeProtectionVectors<a class="headerlink" href="#wipeprotectionvectors" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">wipeProtectionVectors</span><span class="p">(</span><span class="n">pop_id</span><span class="p">,</span> <span class="n">group_id</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Removes all protection sequences from vectors.</p>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>pop_id – ID of population to be modified (String)</p></li>
</ul>
<p><em>Keyword Arguments:</em></p>
<ul class="simple">
<li><p>group_id – ID of group to sample vectors to sample from, if empty, takes
whole population (default empty String; String)</p></li>
</ul>
</section>
<section id="setsetup">
<h4>setSetup<a class="headerlink" href="#setsetup" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">setSetup</span><span class="p">(</span><span class="n">pop_id</span><span class="p">,</span> <span class="n">setup_id</span><span class="p">)</span>
</pre></div>
</div>
<p>Assign parameters stored in Setup object to this population.</p>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>pop_id – ID of population to be modified (String)</p></li>
<li><p>setup_id – ID of setup to be assigned (String)</p></li>
</ul>
</section>
<section id="custommodelfunction">
<h4>customModelFunction<a class="headerlink" href="#custommodelfunction" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">customModelFunction</span><span class="p">(</span><span class="n">function</span><span class="p">)</span>
</pre></div>
</div>
<p>Returns output of given function, passing this model as a parameter.</p>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>function – function to be evaluated; must take a Model object as the
only parameter (function)</p></li>
</ul>
<p><em>Returns:</em></p>
<ul class="simple">
<li><p>Output of function passed as parameter</p></li>
</ul>
</section>
<section id="peaklandscape">
<h4>peakLandscape<a class="headerlink" href="#peaklandscape" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">peakLandscape</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">peak_genome</span><span class="p">,</span> <span class="n">min_value</span><span class="p">)</span>
</pre></div>
</div>
<p>Evaluate a genome’s numerical phenotype by decreasing with distance from optimal
seq.</p>
<p>Originally meant as a purifying selection fitness function based on exponential
decay of fitness as genomes move away from the optimal sequence. Distance is
measured as percent Hamming distance from an optimal genome sequence.</p>
<p>Can be used to evaluate mortality as well as transmissibility.</p>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>genome – the genome to be evaluated (String)</p></li>
<li><p>peak_genome – the genome sequence to measure distance against, has
value of 1 (String)</p></li>
<li><p>min_value – minimum value at maximum distance from optimal
genome (number &gt; 0)</p></li>
</ul>
<p><em>Returns:</em></p>
<ul class="simple">
<li><p>value of genome (number)</p></li>
</ul>
</section>
<section id="valleylandscape">
<h4>valleyLandscape<a class="headerlink" href="#valleylandscape" title="Link to this heading"></a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">valleyLandscape</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">worst_genome</span><span class="p">,</span> <span class="n">min_fitness</span><span class="p">)</span>
</pre></div>
</div>
<p>Evaluate a genome’s numerical phenotype by increasing with distance from worst
seq.</p>
<p>Originally meant as a disruptive selection fitness function based on exponential
decay of fitness as genomes move closer to the worst possible sequence. Distance
is measured as percent Hamming distance from the worst possible genome
sequence.</p>
<p>Can be used to evaluate mortality as well as transmissibility.</p>
<p><em>Arguments:</em></p>
<ul class="simple">
<li><p>genome – the genome to be evaluated (String)</p></li>
<li><p>valley_genome – the genome sequence to measure distance against, has
value of min_value (String)</p></li>
<li><p>min_value – fitness value of worst possible genome (number &gt; 0)</p></li>
</ul>
<p><em>Returns:</em></p>
<ul class="simple">
<li><p>value of genome (number)</p></li>
</ul>
</section>
</section>
</section>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Pablo Cárdenas.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>